---
title: "Untitled"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library("tximport")
library("GenomicFeatures")
library("GenomicAlignments")
library("Rsamtools")
library("readr")
library("DESeq2")

library("magrittr")
library("dplyr")
library("ggplot2")
library("ggridges")
library("VennDiagram")
library("AnnotationHub")
library("goseq")
library("RColorBrewer")
library("pheatmap")
library("BiocParallel")
library("Rsamtools")
#source("bcor_hESC_helperfunctions.R")
source("helper_functions2c.R")

load("../hESC/hgnc.rdata")

ts<-format(Sys.time(), "%a_%b_%d_%Y_%H%M")
cbPalette <- c("#E69F00","#999999","grey", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```



# Tximport -> DESeq2
```{r}

tx2gene<-read.delim("../hESC/salmon/bcor1a_gcbias/quant.sf",stringsAsFactors = F)
tx2gene<-tx2gene[,1,drop=F]
colnames(tx2gene)<-"TXNAME"
tx2gene$GENEID<-sapply(strsplit(tx2gene$TXNAME,"\\|"),function(x) x[2]) %>% substr(.,1,15)
head(tx2gene)

samples<-list.files("../hESC/salmon") %>% grep("gcbias",.,value=T) %>% grep("trim",.,invert = T,value = T)
files <- file.path( "../hESC/salmon", samples, "quant.sf")
names(files) <- samples %>%  gsub("_gcbias","",.)
all(file.exists(files))

txi<-tximport(files,type="salmon",tx2gene = tx2gene)
head(txi$counts)
#sampleTable <- data.frame(sample = factor(c("BCOR","BCOR","H1","H1P","H1","H1P","KDM2B","KDM2B","RAB","RAB","WT"),
#                                          levels=c("H1P","BCOR","H1","RAB","WT","KDM2B")))
sampleTable<-data.frame(sample=factor(c(rep("BCOR",9),"H1","H1P","H1","H1P","KDM2B","KDM2B",rep("RAB",9),"WT"),
                        levels=c("H1P","BCOR","H1","RAB","WT","KDM2B")))
rownames(sampleTable) <- colnames(txi$counts)
dds <- DESeqDataSetFromTximport(txi, sampleTable, ~sample)
dds<-estimateSizeFactors(dds)
colData(dds)$time<-c(0:6,rep(10,8),0:6,10,10,0)

plotPCA(normTransform(dds),intgroup=c("sample","time"))+ theme_bw() +
  scale_color_manual(values = c(cbPalette[c(rep(2,7),8,rep(1,2),5,rep(6,7),8,9)]))


  
```

#Make tabels of FPKM and Counts

```{r}
f<-fpkm(dds,robust=T)
f["ENSG00000163508",] #EOMES
f["ENSG00000125798",] #FOXA2
f["ENSG00000164458",] #T
f["ENSG00000164736",] #SOX17
f["ENSG00000141448",] #GATA6
f["ENSG00000185155",] #MIXL1
f["ENSG00000136574",] #GATA4

counts<-counts(dds,normalized=T)
counts["ENSG00000163508",] #EOMES
counts["ENSG00000125798",] #FOXA2
counts["ENSG00000164458",] #T
counts["ENSG00000164736",] #SOX17
counts["ENSG00000141448",] #GATA6


gg_plotCounts("GSC")
gg_plotCounts("EN2")
gg_plotCounts("NEUROD1")
gg_plotCounts("MIXL1")
gg_plotCounts("EOMES")
gg_plotCounts("GATA4")

temp<-as.data.frame(f)
colnames(temp)<-c("BCOR_0","BCOR_1","BCOR_2","BCOR_3","BCOR_4","BCOR_5","BCOR_6",
                  "H1 shBCOR Rep1","H1 shBCOR Rep2","H1_shCntrl2 Rep1","H1 shCntrl1 Rep1","H1_shCntrl2 Rep2","H1 shCntrl1 Rep2","KDM2B_KO1","KDM2B_KO2",
                  "R1B_0","R1B_1","R1B_2","R1B_3","R1B_4","R1B_5","R1B_6","H1_shRAB Rep1","H1_shRAB Rep2","WT_H1")
summary(idx<-match(rownames(temp),hgnc$ensembl_gene_id))
temp$`HGNC Symbol`<-hgnc[idx,"hgnc_symbol"]
temp<-temp[,c("HGNC Symbol","H1 shBCOR Rep1","H1 shBCOR Rep2","H1 shCntrl1 Rep1","H1 shCntrl1 Rep2","H1_shRAB Rep1","H1_shRAB Rep2","H1_shCntrl2 Rep1","H1_shCntrl2 Rep2","WT_H1",
"KDM2B_KO1","KDM2B_KO2","R1B_0","R1B_1","R1B_2","R1B_3","R1B_4","R1B_5","R1B_6","BCOR_0","BCOR_1","BCOR_2","BCOR_3","BCOR_4","BCOR_5","BCOR_6")]
write.csv(temp,file="BCOR_hESC_normalized_FPKM.csv",quote=F,row.names=T)
```

Differential Expression
```{r}
dds_sh<-dds[,colData(dds)$time==10 | (colData(dds)$time==0 & colData(dds)$sample=="WT")]
dds_sh<-DESeq(dds_sh)
alpha<-0.05
lfc_thres<-0.0

summary(res_bcorsh<-results(dds,contrast=c("sample","BCOR","H1P"),lfcThreshold=0,alpha=alpha))
#plot(res_bcorsh$log2FoldChange,-1*log10(res_bcorsh$padj),cex=0.5,pch=16,xlim=c(-7,12),ylim=c(0,225),main="BCOR")
#res_bcorsh$hgnc<-hgnc[match(rownames(res_bcorsh),hgnc$ensembl_gene_id),"hgnc_symbol"]

summary(res_rabsh<-results(dds,contrast=c("sample","RAB","H1"),lfcThreshold=0,alpha=alpha))
#plot(res_rabsh$log2FoldChange,-1*log10(res_rabsh$padj),cex=0.5,pch=16,xlim=c(-7,12),ylim=c(0,225),main="RAB")

summary(res_kdm2b<-results(dds,contrast=c("sample","KDM2B","WT"),lfcThreshold=0,alpha=alpha))
#plot(res_kdm2b$log2FoldChange,-1*log10(res_kdm2b$padj),cex=0.5,pch=16,xlim=c(-7,12),ylim=c(0,225),main="KDM2B")

#combine for export to CSV 
colnames(res_bcorsh)<-paste0(colnames(res_bcorsh),"_bcorsh")
colnames(res_rabsh)<-paste0(colnames(res_rabsh),"_rabsh")
colnames(res_kdm2b)<-paste0(colnames(res_kdm2b),"_k2bKO")

all.equal(rownames(res_kdm2b),rownames(res_bcorsh))
all.equal(rownames(res_kdm2b),rownames(res_rabsh))

res<-cbind(res_bcorsh,res_rabsh,res_kdm2b)
summary(idx<-match(rownames(res),hgnc$ensembl_gene_id))
res$hgnc<-hgnc[idx,"hgnc_symbol"]

summary(idx<-match(rownames(res),x$ensembl))
res$class<-x[idx,"class"]
res<-res[,c(19,20,1:6,8:12,14:18)]
head(res)
#write.csv(res,file="deseq2_moderated_log2FC_data_090517.csv",quote=F,row.names = T)

#f_sh<-fpkm(dds_shRNA,robust = TRUE)
#par(mfrow=c(1,1))
```

#Venn Diagram
```{r}
summary(degs_bcor<-rownames(subset(res_bcorsh,padj<alpha & abs(log2FoldChange) > lfc_thres)))
summary(degs_rab<-rownames(subset(res_rabsh,padj<alpha & abs(log2FoldChange) > lfc_thres)))
summary(degs_kdm2b<-rownames(subset(res_kdm2b,padj<alpha & abs(log2FoldChange) > lfc_thres)))

overrideTriple<-T
grid.newpage();
vennplot <- draw.triple.venn(length(degs_bcor),
                             length(degs_rab),
                             length(degs_kdm2b),
                             sum(degs_bcor %in% degs_rab), #n12
                             sum(degs_rab %in% degs_kdm2b), #n23
                             sum(degs_bcor %in% degs_kdm2b), #n13
                             sum(degs_bcor[degs_bcor %in% degs_rab] %in% degs_kdm2b), #n123
                             category=c("BCOR KD","RAB KD","KDM2B KO"),
                             cat.col=c("#A11DAB",cbPalette[1],cbPalette[5]),
                             fill=c("#A11DAB",cbPalette[1],cbPalette[5]),
                             scaled=T,euler.d=T)

```

# Classify TSS's
```{r make_tss_prom,eval=F}
txdb<-makeTxDbFromGFF("../gencode.v27.basic.annotation.gff3",format = "gff3")
tss_prom<-trim(promoters(txdb,upstream=1000,downstream=1000))

tss_prom$tx_name<-substr(tss_prom$tx_name,1,15)
names(tss_prom)<-tss_prom$tx_name
sum(duplicated(names(tss_prom)))
summary(idx<-match(substr(names(tss_prom),1,15),hgnc$ensembl_transcript_id))
tss_prom$hgnc<-hgnc[idx,"hgnc_symbol"]
tss_prom$ensembl<-hgnc[idx,"ensembl_gene_id"]
mcols(tss_prom)<-mcols(tss_prom)[,c("hgnc","ensembl")]

save(tss_prom,file="tss_prom.rdata")
```

```{r count_tss,eval=F}
load("tss_prom.rdata")
register(MulticoreParam(workers=16))

fls<- wildtype_fileset@filename %>% grep("ENCFF",.,invert=TRUE, value = TRUE)
fls<-c(fls,"H1_egs_input_P36_Index6_S10.fastq.bam","H1_k2bko_input_P36_Index8_S12.fastq.bam")
bamlst<-BamFileList(fls,yieldSize = 1e7)
seqlevelsStyle(tss_prom)<-"NCBI"

#inter.feature=FALSE will allow reads to count for multiple features
system.time(tss_counts_ncbi<-summarizeOverlaps(tss_prom,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(tss_counts_ncbi)$counts,2,sum)
system.time(tsswide_counts_ncbi<-summarizeOverlaps(tss_prom+9000,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(tsswide_counts_ncbi)$counts,2,sum)

fls<- wildtype_fileset@filename %>% grep("ENCFF",.,invert=FALSE, value = TRUE)
bamlst<-BamFileList(fls,yieldSize = 1e7)
seqlevelsStyle(tss_prom)<-"UCSC"
system.time(tss_counts_ucsc<-summarizeOverlaps(tss_prom,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(tss_counts_ucsc)$counts,2,sum)
system.time(tsswide_counts_ucsc<-summarizeOverlaps(tss_prom+9000,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(tsswide_counts_ucsc)$counts,2,sum)

save(tss_counts_ncbi,tss_counts_ucsc,tsswide_counts_ncbi,tsswide_counts_ucsc,file="tss_counts.rdata")
```

Classify TSS_counts
```{r classify_tss,eval=F}
load("tss_counts.rdata")
load("tss_prom.rdata")

tss_counts<-cbind(assays(tss_counts_ncbi)$counts,assays(tss_counts_ucsc)$counts)
colnames(tss_counts)<-c("BCOR","KDM2B","RNF2","PCGF1","RYBP","CGI","H2AUb","H3K36me2","Ser5p","Ser7p","INPUT_EGS","INPUT","SUZ12","H3K27me3","CBX2","DNAse1","H3K4me3","H3K79me2")
tss_counts<-tss_counts[,c("INPUT_EGS","INPUT","BCOR","KDM2B","RNF2","PCGF1","RYBP","CGI","H2AUb","H3K36me2","SUZ12","H3K27me3","CBX2","DNAse1","H3K4me3","H3K79me2","Ser5p","Ser7p")]

tsswide_counts<-cbind(assays(tsswide_counts_ncbi)$counts,assays(tsswide_counts_ucsc)$counts)
colnames(tsswide_counts)<-c("BCOR","KDM2B","RNF2","PCGF1","RYBP","CGI","H2AUb","H3K36me2","Ser5p","Ser7p","INPUT_EGS","INPUT","SUZ12","H3K27me3","CBX2","DNAse1","H3K4me3","H3K79me2")
tsswide_counts<-tsswide_counts[,c("INPUT_EGS","INPUT","BCOR","KDM2B","RNF2","PCGF1","RYBP","CGI","H2AUb","H3K36me2","SUZ12","H3K27me3","CBX2","DNAse1","H3K4me3","H3K79me2","Ser5p","Ser7p")]

round(data.frame(narrow=colSums(tss_counts),wide=colSums(tsswide_counts))/1e6,2)

colnames(tsswide_counts)<-paste0(colnames(tsswide_counts),"_wide")

tss_counts<-as.data.frame(cbind(tss_counts,tsswide_counts))

#convert to GRanges object
all.equal(names(tss_prom),rownames(tss_counts))
tss_prom$tx_name<-names(tss_prom)
tss_prom$wls<-wls(tss_prom)
names(tss_prom)<-1:length(tss_prom)
rownames(tss_counts)<-1:nrow(tss_counts)
all.equal(names(tss_prom),rownames(tss_counts))
mcols(tss_prom)<-cbind(mcols(tss_prom),tss_counts)

#filter
length(tss_prom<-tss_prom[!duplicated(tss_prom$wls)]) #remove redundant regions

tss_prom<-tss_prom[with(tss_prom,order(-Ser5p))]
length(tss_prom<-tss_prom[!duplicated(tss_prom$ensembl)] )  #select 1 tx for each gene with highest Ser5p

length(tss_prom<-tss_prom[tss_prom$INPUT_EGS_wide < 400 & tss_prom$INPUT_wide < 400] )  #remove aberrant inputs

#summary(idx<-match(tss_prom$tx_name,hgnc$ensembl_transcript_id))
#tss_prom$biotype<-hgnc[idx,"transcript_biotype"]
#x<-data.frame(table(tss_prom$biotype))
#x<-x[with(x,order(-Freq)),]
#rownames(x)<-1:nrow(x)
#length(tss_prom)

#colnames(mcols(tss_prom))
x<-mcols(tss_prom)[,5:22]
kx<-kmeans(x,3)
tss_prom$class<-kx$cluster
table(tss_prom$class)
#boxplot(tss_prom$BCOR~tss_prom$class,main="BCOR +/- 1kb")
#boxplot(tss_prom$BCOR_wide~tss_prom$class,main="BCOR +/- 10kb")
#boxplot(tss_prom$H3K27me3~tss_prom$class,main="H3K27me3 +/- 1kb")
#boxplot(tss_prom$H3K4me3~tss_prom$class,main="H3K4me3 +/- 1kb")
#boxplot(tss_prom$Ser7p~tss_prom$class,main="Ser7p +/- 1kb")
#boxplot(tss_prom$Ser5p~tss_prom$class,main="Ser5p +/- 1kb")

a<-c(mean(tss_prom[tss_prom$class==1]$BCOR),mean(tss_prom[tss_prom$class==2]$BCOR),mean(tss_prom[tss_prom$class==3]$BCOR))

length(tss_strong<-tss_prom[tss_prom$class==which.max(a)])
tss_strong<-tss_strong[with(tss_strong,order(-BCOR))]
tss_strong$class<-"strong"
length(tss_weak<-tss_prom[tss_prom$class==which(a==median(a))])
tss_weak<-tss_weak[with(tss_weak,order(-BCOR))]
tss_weak$class<-"weak"
length(tss_nobinding<-tss_prom[tss_prom$class==which.min(a)])
tss_nobinding<-tss_nobinding[with(tss_nobinding,order(-BCOR))]
tss_nobinding$class<-"nobinding"
length(tss_classified<-c(tss_strong,tss_weak,tss_nobinding))
barplot(tss_classified$Ser5p_wide)
barplot(tss_classified$H3K4me3)
table(tss_classified$class)

seqlevelsStyle(tss_classified)<-"NCBI"
#remove regions that fall off the edge of chromosomes
x<-tss_classified+50000
seqinfo(x)<-seqinfo(BSgenome.Hsapiens.NCBI.GRCh38)
sum(width(trim(x))!=width(x))
length(tss_classified<-tss_classified[width(trim(x))==width(x)])

sum(duplicated(tss_classified$wls))

save(tss_classified,file="tss_classified_081817.rdata")
```

Assign FPKM values to TSS classification
```{r}
load("../hESC/tss_classified_081817.rdata")
x<-mcols(tss_classified)
res_bcorsh$class<-x[match(rownames(res_bcorsh),x$ensembl),"class"]
res_rabsh$class<-x[match(rownames(res_rabsh),x$ensembl),"class"]
res_kdm2b$class<-x[match(rownames(res_kdm2b),x$ensembl),"class"]

res_bcorsh$wt_fpkm<-f[match(rownames(res_bcorsh),rownames(f)),"wt"]
res_rabsh$wt_fpkm<-f[match(rownames(res_rabsh),rownames(f)),"wt"]
res_kdm2b$wt_fpkm<-f[match(rownames(res_kdm2b),rownames(f)),"wt"]

df <- rbind( data.frame(res_bcorsh[,c("baseMean","log2FoldChange","padj","class","wt_fpkm")],factor="BCOR"),
             data.frame(res_rabsh[,c("baseMean","log2FoldChange","padj","class","wt_fpkm")],factor="RAB"),
             data.frame(res_kdm2b[,c("baseMean","log2FoldChange","padj","class","wt_fpkm")],factor="KDM2B"))

table(df$class,useNA="ifany")
#df %>% 
#  dplyr::filter(!(is.na(log2FoldChange) | is.na(padj) | is.na(class))) %>% 
#  mutate(plog= -1*log10(padj)) %>% 
#  mutate(class= factor(class,levels=c("strong","weak","nobinding"))) %>% 
#  mutate(factor= factor(factor,levels=c("BCOR","RAB","KDM2B"))) %>% 
#  mutate(log2fpkm=log2(wt_fpkm+1)) %>% 
#  ggplot(aes(x=log2FoldChange,y=plog,color=log2fpkm)) + geom_point() + 
#  facet_grid( class ~ factor) +theme_bw() +scale_color_gradientn(colors=rainbow(10),name="Log2(WT FPKM + 1)")

```




Figure 2f
```{r add_fpkm,eval=T}
#f_sh["ENSG00000163508",]


summary(idx<-match(tss_classified$ensembl,rownames(f)))
tss_classified$H1P_fpkm<-(f[idx,"h1p1a"]+f[idx,"h1p2a"])/2
tss_classified$BCOR_fpkm<-(f[idx,"bcor1a"]+f[idx,"bcor2a"])/2

tss_classified$H1_H1P_fpkm<-(f[idx,"H1P1"]+f[idx,"H1P2"])/2
tss_classified$H1_RAB_fpkm<-(f[idx,"RAB1"]+f[idx,"RAB2"])/2

tss_classified$WT_fpkm<-f[idx,"wt"]
tss_classified$KDM2B_fpkm<-(f[idx,"kdm2b1"]+f[idx,"kdm2b2"])/2

#write.csv(mcols(tss_classified),file="chip_enrichment_and_fpkm_data_090517.csv",quote=F,row.names = F)

tx2gene$symbol<-sapply(strsplit(tx2gene[,"TXNAME"],"\\|"), function(x) x[6])
tx2gene$biotype<-factor(sapply(strsplit(tx2gene[,"TXNAME"],"\\|"), function(x) x[8]))
tx2gene$biotype<-factor(tx2gene$biotype,levels(tx2gene$biotype)[c(24,12,14,22,1:11,13,15:21,23,25:49)])
levels(tx2gene$biotype)

rev(sort(table(as.character(tx2gene$biotype))))

tx2gene<-tx2gene[with(tx2gene,order(biotype)),]


dim(f2<-as.data.frame(round(f,3)))
dim(f2<-f2[rowSums(f2)> 0.0,])
summary(idx<-match(rownames(f2),tx2gene$GENEID))
f2$symbol<-tx2gene[idx,"symbol"]
f2$biotype<-as.character(tx2gene[idx,"biotype"])
f2<-f2[,c("symbol","biotype","h1p1a","h1p2a","bcor1a","bcor2a","H1P1","H1P2","RAB1","RAB2","wt","kdm2b1","kdm2b2")]

genes_of_interest<-c("EOMES","T","FOXA2","GATA6","GSX2","SOX1","VAX2","GSC","NEUROD1","EN2","SOX17","MIXL1","YAF2","RYBP","CBX2","BCOR","BCORL1")
f2[f2$symbol %in% genes_of_interest,]

write.csv(f2,paste0("fpkm_table_per_sample_",ts,".csv"),quote=F)
R.utils::gzip(paste0("fpkm_table_per_sample_",ts,".csv"))

#from summarized data
(idx<-match(genes_of_interest,tss_classified$hgnc))
temp<-data.frame(symbol=genes_of_interest,round(as.data.frame(mcols(tss_classified)[idx,grep("fpkm",colnames(mcols(tss_classified)))]),3))
temp


g1<-mcols(tss_classified) %>% as.data.frame() %>% 
  dplyr::select(H1P_fpkm,BCOR_fpkm,class) %>% 
  dplyr::filter(!(is.na(H1P_fpkm) | is.na(BCOR_fpkm) | is.na(class))) %>% 
  mutate(H1P= log2(H1P_fpkm+1)) %>% 
  mutate(BCOR= log2(BCOR_fpkm+1)) %>% 
  dplyr::select(H1P,BCOR,class) %>% 
  tidyr::gather(sh,fpkm,na.rm=T,factor_key=T,-class) %>% 
  mutate(class= factor(class,levels=c("strong","weak","nobinding"))) %>%  
  ggplot(aes(x=sh,y=fpkm,fill=sh)) +geom_boxplot() + facet_grid(. ~ class) +
  scale_fill_manual(values=c("#A2AED9","#EF3A38"),name="shRNA",labels=c("Control KD","BCOR KD"))+
  theme_bw()+ ylab("Expression - Log2( FPKM + 1)") +xlab("") +
  theme(strip.background = element_blank(), 
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
ggsave("test.svg",device=svglite::svglite,height=6,width=5)

read_pptx() %>% 
  add_slide(layout = "Title and Content", master = "Office Theme") %>% 
  ph_with_vg(code = print(g1), type = "body") %>% 
  print(target = "Figure_2F.pptx") %>% 
  invisible()


x1<-mcols(tss_classified) %>% as.data.frame() %>% 
  dplyr::filter(class=="strong") %>% 
  dplyr::select(H1P_fpkm,BCOR_fpkm) %>% 
# dplyr::filter(H1P_fpkm > 2.5 | BCOR_fpkm > 2.5) %>% 
 #dplyr::filter(H1P_fpkm > 2.5 | BCOR_fpkm > 2.5) %>% 
  mutate(H1P= log2(H1P_fpkm)) %>% 
  mutate(BCOR= log2(BCOR_fpkm)) %>% 
  dplyr::select(H1P,BCOR) 
plot(x1)
abline(a=0,b=1,col="red")

#%$%  ks.test(H1P,BCOR)$p.value

boxplot(x1)
ks.test(x1$H1P,x1$BCOR)$p.value
t.test(x1$H1P,x1$BCOR,paired=F)$p.value

  wilcox.test(x1$H1P_fpkm,x1$BCOR_fpkm,paired=T)$p.value




%>% 
  mutate(H1P= log2(H1P_fpkm+1)) %>% 
  mutate(BCOR= log2(BCOR_fpkm+1)) %>% 
  dplyr::select(H1P,BCOR) 

wilcox.test(x1$H1P,x1$BCOR,paired=T)


mcols(tss_classified) %>% as.data.frame() %>% 
  dplyr::select(class:KDM2B_fpkm) %>% 
  dplyr::filter(!(is.na(H1P_fpkm) | is.na(BCOR_fpkm) | is.na(class))) %>% 
 # mutate(H1P= log2(H1P_fpkm+1)) %>% 
 # mutate(BCOR= log2(BCOR_fpkm+1)) %>% 
#  dplyr::select(H1P,BCOR,class) %>% 
  tidyr::gather(sh,fpkm,na.rm=T,factor_key=T,-class) %>% 
  mutate(class= factor(class,levels=c("strong","weak","nobinding"))) %>%  
 mutate(fpkm=log2(fpkm + 1)) %>% 
   ggplot(aes(x=sh,y=fpkm,fill=sh)) +geom_boxplot() + facet_grid(. ~ class) +
  scale_fill_manual(values = c("#BFBFBF","#A11DAB",cbPalette[c(2,1,6,5)]),name="shRNA")+
  theme_bw()+ ylab("Expression - Log2( FPKM + 1)") +xlab("") +
  theme(strip.background = element_blank(), 
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

write.csv(mcols(tss_classified),file=paste0("chip_enrichment_and_fpkm_data_",ts,".csv",quote=F,row.names = F)
```

```{r}

idx<-match(tss_classified$ensembl,rownames(f))

f[idx,c(4,6,1,2)] %>% as.data.frame() 

g1<-mcols(tss_classified) %>% as.data.frame() %>% 
  dplyr::filter(class=="weak" | class=="strong") %>% 
  dplyr::select(H1P_fpkm,BCOR_fpkm,class) %>% 
  dplyr::mutate(pos=1:n() ) %>% 
  dplyr::filter(!(is.na(BCOR_fpkm) | is.na(H1P_fpkm)) ) %>% 
  tidyr::gather(sample,fpkm,-class:-pos) %>%
  dplyr::mutate(fpkm=log2(fpkm+1)) %>% 
  ggplot(aes(x=pos,y=fpkm,group=sample,colour=sample)) + 
 # stat_smooth(subset(class="strong"), method = "auto") +
  stat_smooth(method="auto") + 
 # geom_point(size=0.2)+
  theme_bw() + coord_flip() + scale_colour_manual(values=c("blue","red")) + 
  xlim(c(16894,1)) + ylab("Log2(FPKM+1)") + 
  xlab("Position in Carpet Plot from Top to Bottom")

g2<-mcols(tss_classified) %>% as.data.frame() %>% 
  dplyr::filter(class=="weak" | class=="strong") %>% 
  dplyr::select(H1_H1P_fpkm,H1_RAB_fpkm,class) %>% 
  dplyr::mutate(pos=1:n() ) %>% 
  dplyr::filter(!(is.na(H1_H1P_fpkm) | is.na(H1_RAB_fpkm)) ) %>% 
  tidyr::gather(sample,fpkm,-class:-pos) %>%
  dplyr::mutate(fpkm=log2(fpkm+1)) %>% 
  ggplot(aes(x=pos,y=fpkm,group=sample,colour=sample)) + 
 # stat_smooth(subset(class="strong"), method = "auto") +
  stat_smooth(method="auto") + 
 # geom_point(size=0.2)+
  theme_bw() + coord_flip() + scale_colour_manual(values=c("blue","red")) + 
  xlim(c(16894,1)) + ylab("Log2(FPKM+1)") + 
  xlab("Position in Carpet Plot from Top to Bottom")

g3<-mcols(tss_classified) %>% as.data.frame() %>% 
  dplyr::filter(class=="weak" | class=="strong") %>% 
  dplyr::select(WT_fpkm,KDM2B_fpkm,class) %>% 
  dplyr::mutate(pos=1:n() ) %>% 
  dplyr::filter(!(is.na(WT_fpkm) | is.na(KDM2B_fpkm)) ) %>% 
  tidyr::gather(sample,fpkm,-class:-pos) %>%
  dplyr::mutate(fpkm=log2(fpkm+1)) %>% 
  ggplot(aes(x=pos,y=fpkm,group=sample,colour=sample)) + 
 # stat_smooth(subset(class="strong"), method = "auto") +
  stat_smooth(method="auto") + 
 # geom_point(size=0.2)+
  theme_bw() + coord_flip() + scale_colour_manual(values=c("blue","red")) + 
  xlim(c(16894,1)) + ylab("Log2(FPKM+1)") + 
  xlab("Position in Carpet Plot from Top to Bottom")

```


Download Files for GO Analysis
```{r}
#download.file("http://geneontology.org/gene-associations/goa_human.gaf.gz",destfile = "goa_human.gaf.gz")
#R.utils::gunzip("goa_human.gaf.gz")
#download.file("http://purl.obolibrary.org/obo/go.obo",destfile = "go.obo")

goa_human<-readr::read_delim("../hESC/H1/goa_human.gaf",delim="\t",comment="!",col_names=c("DB","DB Object ID","Symbol","Qualifier","GO ID","DB:eEference","Evidence Code",
                    "WithorFrom","Aspect","DB Object Name","DB Object Syn","DB Object Type","Taxon","Date","Assigned By",
                    "Annotation Extension","Gene Product Form ID"))
table(goa_human$`Evidence Code`)


t1<-readLines("../hESC/H1/go.obo",n=-1L)
go_terms<-list()


for (i in seq_along(1:length(t1))) {
    if (t1[i]=="[Term]") {
      gt<-substr(t1[i+1],5,14)
      name<-gsub("name: ","",t1[i+2])
      if (t1[i+3]=="namespace: biological_process") { 
        #print (paste0(gt,":  ",name)) 
        go_terms[[gt]]<-name
      }
      }
  }
go_terms<-unlist(go_terms)
length(go_terms)

#subset associations to BP
sum(goa_human$`GO ID` %in% names(go_terms))
dim(goa_human<-goa_human[goa_human$`GO ID` %in% names(go_terms),])

#add ensemblid for each row
ah <- AnnotationHub()
query(ah, "EnsDb")
ens88<-query(ah, pattern = c("Homo Sapiens", "EnsDb", 88))[[1]]

temp<-AnnotationDbi::select(ens88,keys=unique(goa_human$`DB Object ID`),keytype="UNIPROTID",columns=c("GENEID","SYMBOL"))
summary(idx<-match(goa_human$`DB Object ID`,temp$UNIPROTID))
goa_human$ensembl<-temp[idx,"GENEID"]
goa_human$symbol<-temp[idx,"SYMBOL"]

save(goa_human,go_terms,file="GO_data.rdata")

#subset GO terms to genes expressed in tss_classified
length(expressed_genes<-tss_classified$ensembl)

dim(goa_human<-goa_human[goa_human$ensembl %in% expressed_genes,])

#reduce to term and gene
dim(goa_human<-goa_human[!is.na(goa_human$ensembl),c("ensembl","GO ID")])

#remove duplicated entries (possibly from multiple lines of evidence)
dim(goa_human<-goa_human[!duplicated(goa_human),])

#shorted expressed genes to just those that we have ontology information for.
length(expressed_genes<-expressed_genes[expressed_genes %in% unique(goa_human$ensembl)])

#make a list
goa_human.list<-split(goa_human$`GO ID`,goa_human$ensembl)

#test
go_terms[goa_human.list[[temp[grep("FOXA2",temp$SYMBOL),"GENEID"]]]]
go_terms[goa_human.list[[temp[grep("^BCOR$",temp$SYMBOL),"GENEID"]]]]
go_terms[goa_human.list[[temp[grep("^DMRT1$",temp$SYMBOL),"GENEID"]]]]

goa_human[grep("GO:0050911",goa_human$`GO ID`),]$ensembl
AnnotationDbi::select(ens88,keys=unique(goa_human[grep("GO:0050911",goa_human$`GO ID`),]$ensembl),keytype="GENEID",columns=c("SYMBOL"))

#Get bias data
bd<-sum(width(reduce(exonsBy(ens88,by="gene"))))
bd<-bd[expressed_genes]
bd[temp[grep("^BCOR$",temp$SYMBOL),"GENEID"]]
```

```{r}
#run goseq on 3 categories
sum(degs<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$class=="strong"]))
names(degs)<-expressed_genes
table(degs)

pwf.strong<-nullp(degs,bias.data=bd)

#weak
sum(degs<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$class=="weak"]))
names(degs)<-expressed_genes
table(degs)

pwf.weak<-nullp(degs,bias.data=bd)

#nobinding
sum(degs<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$class=="nobinding"]))
names(degs)<-expressed_genes
table(degs)
pwf.nobinding<-nullp(degs,bias.data=bd)

GO.hyper.strong.goa<-goseq(pwf.strong,gene2cat=goa_human.list,method = "Hypergeometric")
GO.hyper.weak.goa<-goseq(pwf.weak,gene2cat=goa_human.list,method = "Hypergeometric")
GO.hyper.nobinding.goa<-goseq(pwf.nobinding,gene2cat=goa_human.list,method = "Hypergeometric")


write.csv(GO.hyper.strong.goa[GO.hyper.strong.goa$over_represented_pvalue < 0.05,],file="GO_hyper_strong_goa.csv",quote=F,row.names = F)
write.csv(GO.hyper.weak.goa[GO.hyper.weak.goa$over_represented_pvalue < 0.05,],file="GO_hyper_weak_goa.csv",quote=F,row.names = F)
write.csv(GO.hyper.nobinding.goa[GO.hyper.nobinding.goa$over_represented_pvalue < 0.05,],file="GO_hyper_nobinding_goa.csv",quote=F,row.names = F)
```

```{r}
revigo.strong<-read.csv("revigo_strong_goa.csv",stringsAsFactors = F,header=T)
revigo.strong$class<-"strong"
revigo.weak<-read.csv("revigo_weak_goa.csv",stringsAsFactors = F,header=T)
revigo.weak$class<-"weak"
revigo.nobinding<-read.csv("revigo_nobinding_goa.csv",stringsAsFactors = F,header=T)
revigo.nobinding$class<-"nobinding"

go_plot<-rbind(revigo.strong,revigo.weak,revigo.nobinding) %>%  
  dplyr::select(description,log10.p.value,class) %>% 
  dplyr::mutate(log10.p.value = -1* log10.p.value) %>% 
  dplyr::filter(log10.p.value > 15) %>% 
  dplyr::mutate(class=factor(class,levels=rev(c("strong","weak","nobinding")))) %>% 
  dplyr::arrange(class,log10.p.value) %>% 
  dplyr::mutate(description=factor(description,levels=description)) %>% 
ggplot(aes(x=description,y=log10.p.value,fill=class)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values=rev(c("#993366","#D37BA7","#BFBFBF")),guide=FALSE) +
  coord_flip() + ylab("-Log10( p-value)") +
  theme_bw() +
  theme(strip.background = element_blank(), 
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_blank())

ggsave("go_terms_090517.png",plot=go_plot,dpi=300,width=6,height=6)

```

filesets
```{r}
bcor_fileset<-new("fileset", filename=c( "../H1_wt_BCOR_P36_Index15_S14.fastq.bam","../ubH2A_H1_081816_GRCh38.bam"),
           labels=c("BCOR","UbH2A"))
(bcor_fileset<-countFileset(bcor_fileset))

pol2_fileset<-new("fileset", filename=c( "../hESC/estaras_Ser5P_RNAPII_untreated_SRR1745499.R1_trimmed.fastq.bam","../hESC/estaras_Ser7P_RNAPII_untreated_SRR1745503.R1_trimmed.fastq.bam"),
           labels=c("Ser5P","Ser7p"))
(pol2_fileset<-countFileset(pol2_fileset))

cbx2rybp_fileset<-new("fileset", filename=c( "../H1_CBX2_broad_ENCFF614JXQ.bam","../H1_egs_RYBP_P36_Index13_S8.fastq.bam"),
           labels=c("CBX2","RYBP"))
(cbx2rybp_fileset<-countFileset(cbx2rybp_fileset))

dnase_fileset<-new("fileset", filename=c( "../DNAse1_H1_UW_ENCFF674MNA.bam"),
           labels=c("DNASE1"))
(dnase_fileset<-countFileset(dnase_fileset))

cgi_fileset<-new("fileset", filename=c( "../cgi_hg38_canon_ncbi.bam"),
           labels=c("CGI"))
(cgi_fileset<-countFileset(cgi_fileset))

prc1_fileset<-new("fileset",filename=c("../H1_egs_KDM2B_P36_Index11_S6.fastq.bam",
                                 "../H1_wt_RNF2_P36_Index16_S15.fastq.bam",
                                 "../H1_egs_PCGF1_P36_Index12_S7.fastq.bam"),
                  labels=c("KDM2B","RNF2","PCGF1"))
prc1_fileset<-countFileset(prc1_fileset)

rna_fileset<-new("fileset", filename=c( "../hESC/H1P1A_CGATGT.bam","../hESC/BCOR1A_TGACCA.bam"),
           labels=c("H1P","shBCOR"))
(rna_fileset<-countFileset(rna_fileset))

input_fileset<-new("fileset",filename=c("../H1_k2bko_BCOR_P36_Index17_S16.fastq.bam",
                                        "../H1_k2bko_input_P36_Index8_S12.fastq.bam"),
                  labels=c("BCOR_k2bko","INPUT_k2bko"))
input_fileset<-countFileset(input_fileset)

(ser7p_fileset<-countFileset(new("fileset",filename=c("../estaras_Ser7P_RNAPII_untreated_SRR1745503.R1_trimmed.fastq.bam",
                                        "../H1_k2bko_input_P36_Index8_S12.fastq.bam"),
                  labels=c("CHIP","INPUT"))))



histones_fileset<-new("fileset", filename=c( "../H1_H3K4me3_ENCFF571AXH.bam","../H1_H3K27me3_ENCFF171RVK.bam",
                                             "../H3K36me2_ChIP_zheng_GCCAAT_bothruns.R1_trimmed.fastq.bam",
                                             "../H1_h3k79me2_broad_ENCFF965LBC.bam",
                                             "../ubH2A_H1_081816_GRCh38.bam"),
           labels=c("H3K4me3","H3K27me3","H3K36me2","H3K79me2","H2AK119ub"))
(histones_fileset<-countFileset(histones_fileset))



(histones_fileset<-countFileset(new("fileset", filename=c( "../hESC/H1_H3K27me3_ENCFF171RVK.bam",
                                             "../hESC/H3K36me2_ChIP_zheng_GCCAAT_bothruns.R1_trimmed.fastq.bam"),
           labels=c("H3K27me3","H3K36me2"))))
```


# Summarize counts for the BCOR ChIP-seqs in K2B KO and RNF2-R for stacked histogram plots
```{r summarize_rnf2r_k2bko,eval=F}
load("filesets.rdata")
load("tss_classified_081817.rdata")
register(MulticoreParam(workers=8))

length(bound<-tss_classified[tss_classified$class!="nobinding"])
seqlevelsStyle(bound)<-"NCBI"
bamlst<-BamFileList(rnf2r_fileset@filename,yieldSize = 1e7)

#inter.feature=FALSE will allow reads to count for multiple features
system.time(rnf2r_counts<-summarizeOverlaps(bound,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(rnf2r_counts)$counts,2,sum)
rnf2r_counts<-as.data.frame(assays(rnf2r_counts)$counts)
stopifnot(all.equal(names(bound),rownames(rnf2r_counts)))
mcols(bound)<-cbind(mcols(bound),rnf2r_counts)
bound<-bound[with(bound, order(-H1_BCORminus_P038_Index2_S13.R1_trimmed.fastq.bam))]
save(bound,file="bound_rnf2r_order.rdata")


#rnf2r<-tornado(bound,pad=9000,rnf2r_fileset,ord=3,color="#0052bf",window=100)
#ggsave(file=paste0(ts,"_figure4x_rnf2r_bound.png"),plot=rnf2r,width=9,height=8)

bamlst<-BamFileList(k2bko_fileset@filename,yieldSize = 1e7)

#inter.feature=FALSE will allow reads to count for multiple features
system.time(k2bko_counts<-summarizeOverlaps(bound,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(k2bko_counts)$counts,2,sum)
k2bko_counts<-as.data.frame(assays(k2bko_counts)$counts)
stopifnot(all.equal(names(bound),rownames(k2bko_counts)))
mcols(bound)<-cbind(mcols(bound),k2bko_counts)
bound<-bound[with(bound, order(-H1_k2bko_BCOR_P36_Index17_S16.fastq.bam))]
save(bound,file="bound_k2bko_order.rdata")

tornado(bound[1:10],pad=9000,k2bko_fileset,ord=3,color="#0052bf",window=100)
#k2bko<-tornado(bound,pad=9000,k2bko_fileset,ord=3,color="#0052bf",window=100)
#ggsave(file=paste0(ts,"_figure4x_k2bko_bound.png"),plot=k2bko,width=9,height=8)

```

```{r}

length(bound_strong<-tss_classified[tss_classified$class=="strong"])
seqlevelsStyle(bound_strong)<-"NCBI"
bamlst<-BamFileList(k2bko_fileset@filename,yieldSize = 1e7)

#inter.feature=FALSE will allow reads to count for multiple features
system.time(k2bko_counts<-summarizeOverlaps(bound_strong,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(k2bko_counts)$counts,2,sum)
k2bko_counts<-as.data.frame(assays(k2bko_counts)$counts)

for (i in seq_along(k2bko_fileset@filename)) {
  print(colnum<-which(colnames(k2bko_counts)==basename(k2bko_fileset@filename[i])))
  colnames(k2bko_counts)[colnum]<-paste0(k2bko_fileset@labels[i]," cpm")
  k2bko_counts[,colnum]<-round(1e6*k2bko_counts[,colnum]/as.numeric(k2bko_fileset@count[i]),3)
}

colnames(k2bko_counts)<-c("BCOR_WT_cpm","BCOR_K2BKO_cpm")
stopifnot(all.equal(names(bound_strong),rownames(k2bko_counts)))
mcols(bound_strong)<-cbind(mcols(bound_strong),k2bko_counts)


system.time(k2bko_counts_wide<-summarizeOverlaps(bound_strong+9000,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(k2bko_counts_wide)$counts,2,sum)
k2bko_counts_wide<-as.data.frame(assays(k2bko_counts_wide)$counts)

for (i in seq_along(k2bko_fileset@filename)) {
  print(colnum<-which(colnames(k2bko_counts_wide)==basename(k2bko_fileset@filename[i])))
  colnames(k2bko_counts_wide)[colnum]<-paste0(k2bko_fileset@labels[i]," wide cpm")
  k2bko_counts_wide[,colnum]<-round(1e6*k2bko_counts_wide[,colnum]/as.numeric(k2bko_fileset@count[i]),3)
}


colnames(k2bko_counts_wide)<-c("BCOR_WT_wide_cpm","BCOR_K2BKO_wide_cpm")
stopifnot(all.equal(names(bound_strong),rownames(k2bko_counts_wide)))
mcols(bound_strong)<-cbind(mcols(bound_strong),k2bko_counts_wide)
head(bound_strong)

bound_strong %>% as.data.frame() %>% 
#  dplyr::select(-hgnc,-ensembl) %>% 
  dplyr::select(BCOR_WT_cpm,BCOR_K2BKO_cpm,BCOR_WT_wide_cpm,BCOR_K2BKO_wide_cpm) %>% 
  tidyr::gather(chip,cpm) %>%
  mutate(chip=factor(chip,levels=c("BCOR_WT_cpm","BCOR_K2BKO_cpm","BCOR_WT_wide_cpm","BCOR_K2BKO_wide_cpm"))) %>% 
  mutate(cpm=log2(cpm)) %>%
  ggplot(aes(x=cpm,y=chip,fill=chip)) + geom_density_ridges() + theme_bw() + 
  scale_fill_manual(values=cbPalette[c(4,7,5,6)]) + #facet_grid(class ~ .) +
  xlab("Log2(cpm)") +ylab("") + ggtitle("KDM2B ChIP Distributions +/- RNF2") 
```

#now I need wide counts for RNF2-R dataset, 
"Thu Sep 28 15:47:40 2017"
```{r}

length(bound<-tss_classified[tss_classified$class!="nobinding"])
seqlevelsStyle(bound)<-"NCBI"


(lof_fileset<-countFileset(new("fileset", filename=c("../hESC/H1_egs_input_P36_Index6_S10.fastq.bam",
                                                     "../hESC/H1_k2bko_input_P36_Index8_S12.fastq.bam",
                                                     "../hESC/H1_wt_BCOR_P36_Index15_S14.fastq.bam",
                                                     "../hESC/H1_k2bko_BCOR_P36_Index17_S16.fastq.bam",
                                                     "../hESC/H1_INPUTplus_P038_Index8_S19.R1_trimmed.fastq.bam",
                                                       "../hESC/H1_INPUTminus_P038_Index6_S17.R1_trimmed.fastq.bam",
                                                       "../hESC/H1_BCORplus_P038_Index4_S15.R1_trimmed.fastq.bam",
                                                       "../hESC/H1_BCORminus_P038_Index2_S13.R1_trimmed.fastq.bam",
                                                       "../hESC/H1_KDM2Bplus_P038_Index5_S16.R1_trimmed.fastq.bam",
                                                       "../hESC/H1_KDM2Bminus_P038_Index3_S14.R1_trimmed.fastq.bam"),
labels=c("INPUT_EGS","INPUT_K2BKO","BCOR_WT","BCOR_K2BKO","INPUT_PLUS","INPUT_MINUS","BCOR_PLUS","BCOR_MINUS","KDM2B_PLUS","KDM2B_MINUS"))))

bamlst<-BamFileList(lof_fileset@filename,yieldSize = 1e7)
#inter.feature=FALSE will allow reads to count for multiple features
system.time(bound_counts<-summarizeOverlaps(bound,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(bound_counts)$counts,2,sum)
dim(bound_counts<-as.data.frame(assays(bound_counts)$counts))


for (i in seq_along(lof_fileset@filename)) {
  print(colnum<-which(colnames(bound_counts)==basename(lof_fileset@filename[i])))
  colnames(bound_counts)[colnum]<-paste0(lof_fileset@labels[i],"_cpm")
  bound_counts[,colnum]<-round(1e6*bound_counts[,colnum]/as.numeric(lof_fileset@count[i]),3)
}

stopifnot(all.equal(names(bound),rownames(bound_counts)))
mcols(bound)<-cbind(mcols(bound),bound_counts)

#inter.feature=FALSE will allow reads to count for multiple features
system.time(bound_counts_wide<-summarizeOverlaps(bound+9000,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(bound_counts_wide)$counts,2,sum)
dim(bound_counts_wide<-as.data.frame(assays(bound_counts_wide)$counts))


for (i in seq_along(lof_fileset@filename)) {
  print(colnum<-which(colnames(bound_counts_wide)==basename(lof_fileset@filename[i])))
  colnames(bound_counts_wide)[colnum]<-paste0(lof_fileset@labels[i],"_wide_cpm")
  bound_counts_wide[,colnum]<-round(1e6*bound_counts_wide[,colnum]/as.numeric(lof_fileset@count[i]),3)
}

stopifnot(all.equal(names(bound),rownames(bound_counts_wide)))
mcols(bound)<-cbind(mcols(bound),bound_counts_wide)

save(bound,file="bound_w_counts.rdata")
```

# this section also replaced by quantification of all bound above?
```{r}

#inter.feature=FALSE will allow reads to count for multiple features
#system.time(rnf2r_counts<-summarizeOverlaps(bound_strong,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(rnf2r_counts)$counts,2,sum)
rnf2r_counts<-as.data.frame(assays(rnf2r_counts)$counts)

for (i in seq_along(rnf2r_fileset@filename)) {
  print(colnum<-which(colnames(rnf2r_counts)==basename(rnf2r_fileset@filename[i])))
  colnames(rnf2r_counts)[colnum]<-paste0(rnf2r_fileset@labels[i]," cpm")
  rnf2r_counts[,colnum]<-round(1e6*rnf2r_counts[,colnum]/as.numeric(rnf2r_fileset@count[i]),3)
}

colnames(rnf2r_counts)<-paste0(rnf2r_fileset@labels,"_cpm")
stopifnot(all.equal(names(bound_strong),rownames(rnf2r_counts)))
mcols(bound_strong)<-cbind(mcols(bound_strong),rnf2r_counts)


system.time(rnf2r_counts_wide<-summarizeOverlaps(bound_strong+9000,bamlst,mode = "Union",ignore.strand = TRUE,inter.feature=FALSE))
apply(assays(rnf2r_counts_wide)$counts,2,sum)
rnf2r_counts_wide<-as.data.frame(assays(rnf2r_counts_wide)$counts)

for (i in seq_along(rnf2r_fileset@filename)) {
  print(colnum<-which(colnames(rnf2r_counts_wide)==basename(rnf2r_fileset@filename[i])))
  colnames(rnf2r_counts_wide)[colnum]<-paste0(rnf2r_fileset@labels[i]," wide cpm")
  rnf2r_counts_wide[,colnum]<-round(1e6*rnf2r_counts_wide[,colnum]/as.numeric(rnf2r_fileset@count[i]),3)
}

colnames(rnf2r_counts_wide)<-paste0(rnf2r_fileset@labels,"_wide_cpm")
stopifnot(all.equal(names(bound_strong),rownames(rnf2r_counts_wide)))
mcols(bound_strong)<-cbind(mcols(bound_strong),rnf2r_counts_wide)

head(bound_strong)

bound_strong_rnf2r_sort <-bound_strong %>% as.data.frame() %>% 
  dplyr::filter(class=="strong") %>%
  dplyr::select(hgnc,ensembl,BCOR_WT_wide_cpm,BCOR_K2BKO_wide_cpm,BCOR_PLUS_wide_cpm,
                BCOR_MINUS_wide_cpm,KDM2B_PLUS_wide_cpm, KDM2B_MINUS_wide_cpm) %>% 
#  dplyr::mutate(e=log2(BCOR_K2BKO) - log2(INPUT)) %>% 
  dplyr::mutate(bk_wide=log2(BCOR_K2BKO_wide_cpm) - log2(BCOR_WT_wide_cpm)) %>% 
  dplyr::mutate(br_wide=log2(BCOR_MINUS_wide_cpm) - log2(BCOR_PLUS_wide_cpm)) %>% 
  dplyr::mutate(kr_wide=log2(KDM2B_MINUS_wide_cpm) - log2(KDM2B_PLUS_wide_cpm)) %>% 
#  dplyr::arrange(-BCOR_MINUS_wide_cpm) 
  dplyr::arrange(br_wide) 

  #dplyr::select(ensembl,e,BCOR_K2BKO)

bound_strong_rnf2r_sort$kdm2b_ex<-res_kdm2b[bound_strong_rnf2r_sort$ensembl,"log2FoldChange_k2bKO"]
bound_strong_rnf2r_sort$rab_ex<-res_rabsh[bound_strong_rnf2r_sort$ensembl,"log2FoldChange_rabsh"]
bound_strong_rnf2r_sort$bcor_ex<-res_bcorsh[bound_strong_rnf2r_sort$ensembl,"log2FoldChange_bcorsh"]

#Match classification status
summary(idx<-match(bound_strong_rnf2r_sort$ensembl,tss_classified$ensembl))
bound_strong_rnf2r_sort$deg_bcor<-tss_classified[idx]$deg_bcor
bound_strong_rnf2r_sort$deg_k2bKO<-tss_classified[idx]$deg_k2bKO
bound_strong_rnf2r_sort$deg_rab<-tss_classified[idx]$deg_rab
table(bound_strong_rnf2r_sort$deg_bcor)
table(bound_strong_rnf2r_sort$deg_k2bKO)
table(bound_strong_rnf2r_sort$deg_rab)


bound_strong_rnf2r_sort[grep("ENSG00000152977",bound_strong_rnf2r_sort$ensembl),] #ZIC1
bound_strong_rnf2r_sort[grep("ENSG00000163508",bound_strong_rnf2r_sort$ensembl),] # EOMES
bound_strong_rnf2r_sort[grep("ENSG00000156574",bound_strong_rnf2r_sort$ensembl),] # NODAL

head(bound_strong_rnf2r_sort)

chip_df<- bound_strong_rnf2r_sort[,3:11] %>% 
    mutate(BCOR_WT=log2(BCOR_WT_wide_cpm)) %>% 
    mutate(BCOR_K2BKO=log2(BCOR_K2BKO_wide_cpm)) %>% 
  #
    mutate(BCOR_wRNF2=log2(BCOR_PLUS_wide_cpm)) %>% 
    mutate(BCOR_woRNF2=log2(BCOR_MINUS_wide_cpm)) %>% 
  #
    mutate(KDM2B_wRNF2=log2(KDM2B_PLUS_wide_cpm)) %>% 
    mutate(KDM2B_woRNF2=log2(KDM2B_MINUS_wide_cpm)) %>% 
    select(BCOR_WT,BCOR_K2BKO,BCOR_wRNF2,BCOR_woRNF2,KDM2B_wRNF2,KDM2B_woRNF2,bk_wide,br_wide,kr_wide)


colors2=colorRampPalette(brewer.pal(11,"PRGn"))(200)
#use breaklist to set midpoint to 0
myBreaksList2 <- c(seq(min(chip_df,na.rm=T), 0, length.out=ceiling(length(colors2)/2) + 1), 
              seq(max(chip_df,na.rm=T)/length(colors2), max(chip_df,na.rm=T), length.out=floor(length(colors2)/2)))
pheatmap(chip_df,cluster_rows=F,cluster_cols=F,labels_row="",col=colors2,breaks=myBreaksList2)

#leave off ChIPs
myBreaksList2 <- c(seq(min(chip_df,na.rm=T), 0, length.out=ceiling(length(colors2)/2) + 1), 
              seq(max(chip_df,na.rm=T)/length(colors2), 1, length.out=floor(length(colors2)/2)))

pheatmap(chip_df[,c(8,9,7)],cluster_rows=F,cluster_cols=F,labels_row="",col=colors2,breaks=myBreaksList2)

png(paste0("Figure6x_chip_heatmap_",ts,".png"),width=1,height=8,units="in",res=300)
pheatmap(chip_df[,c(8,9)],cluster_rows=F,cluster_cols=F,labels_row="",col=colors2,breaks=myBreaksList2)
dev.off()
cor(bound_strong_rnf2r_sort$br_wide,bound_strong_rnf2r_sort$bk_wide,method="spearman")

gg1<-data.frame(pos=nrow(bound_strong_rnf2r_sort):1,
           bcorFC_rnf2r=rev(bound_strong_rnf2r_sort$br_wide),
           bcorFC_kdm2bko=rev(bound_strong_rnf2r_sort$bk_wide)) %>% 
  tidyr::gather(ko,y,-pos) %>% 
  ggplot(aes(x=pos,y=y,colour=ko)) +geom_point(size=0.05) + geom_smooth(size=2) + theme_bw() +
  scale_color_manual(name="BCOR ChIP\nLogFC",
        values=c(bcorFC_rnf2r="red",bcorFC_kdm2bko="green2")) +
  theme(  panel.grid.major=element_blank(),
                         panel.grid.minor=element_blank())

ggsave(file=paste0("figure6x_chip_smooth",ts,".pdf"),plot=gg1,width=7,height=5)


colors <- colorRampPalette(c("#A2AED9","snow","#B51F2E"))(200)
#expression_df<- bound_strong_rnf2r_sort[,c("rab_ex","bcor_ex","kdm2b_ex")]
expression_df<-temp2[bound_strong_rnf2r_sort$ensembl, 7:12] #temp2 is defined below
myBreaksList <- c(seq(min(expression_df,na.rm=T), 0, length.out=ceiling(length(colors)/2) + 1), 
              seq(max(expression_df,na.rm=T)/length(colors), max(expression_df,na.rm=T), length.out=floor(length(colors)/2)))

pheatmap(expression_df,cluster_rows=F,cluster_cols=F,labels_row="",col=colors,
         breaks=myBreaksList)

png(paste0("Figure6x_expression_heatmap_",ts,".png"),width=1,height=8,units="in",res=300)
pheatmap(expression_df,cluster_rows=F,cluster_cols=F,labels_row="",col=colors,breaks=myBreaksList)
dev.off()

#(k2bko_fileset<-countFileset(new("fileset", filename=c("../hESC/H1_wt_BCOR_P36_Index15_S14.fastq.bam",
#"../hESC/H1_k2bko_BCOR_P36_Index17_S16.fastq.bam"),
#           labels=c("BCOR WT","BCOR K2BKO"))))

summary(idx<-match(bound_strong_rnf2r_sort$ensembl,tss_classified$ensembl))

(rnf2r_fileset<-countFileset(new("fileset", filename=c("../hESC/H1_BCORplus_P038_Index4_S15.R1_trimmed.fastq.bam",
                                                       "../hESC/H1_BCORminus_P038_Index2_S13.R1_trimmed.fastq.bam",
                                                       "../hESC/H1_KDM2Bplus_P038_Index5_S16.R1_trimmed.fastq.bam",
                                                       "../hESC/H1_KDM2Bminus_P038_Index3_S14.R1_trimmed.fastq.bam"),
labels=c("BCOR_PLUS","BCOR_MINUS","KDM2B_PLUS","KDM2B_MINUS"))))

system.time(rnf2r_tornado<-tornado(tss_classified[idx],pad=9000,rnf2r_fileset,ord=0,color="#00441B",window=100)) #green
ggsave(file=paste0("figure6x_rnf2r_bound_",ts,".png"),plot=rnf2r_tornado+theme(panel.border = element_blank()),width=7,height=8)


```



#don't need this secion anymore?
```{r}
load("bound_rnf2r_order_Fri_Sep_15_2017_1958.rdata")
bound[grep("EOMES",bound$hgnc),]
bound_rnf2r<-mcols(bound)[,c(1:2,41:47)]

for (i in seq_along(rnf2r_fileset@filename)) {
  print(colnum<-which(colnames(bound_rnf2r)==basename(rnf2r_fileset@filename[i])))
  colnames(bound_rnf2r)[colnum]<-paste0(rnf2r_fileset@labels[i]," cpm")
  bound_rnf2r[,colnum]<-round(1e6*bound_rnf2r[,colnum]/as.numeric(rnf2r_fileset@count[i]),3)
}
head(bound_rnf2r)


load("bound_k2bko_order_Fri_Sep_15_2017_1958.rdata")
bound_k2bko<-mcols(bound)[,c(1:2,41:44)]

for (i in seq_along(k2bko_fileset@filename)) {
  print(colnum<-which(colnames(bound_k2bko)==basename(k2bko_fileset@filename[i])))
  colnames(bound_k2bko)[colnum]<-paste0(k2bko_fileset@labels[i]," cpm")
  bound_k2bko[,colnum]<-round(1e6*bound_k2bko[,colnum]/as.numeric(k2bko_fileset@count[i]),3)
}
head(bound_k2bko)

bound<-bound_rnf2r
summary(idx<-match(bound$ensembl,bound_k2bko$ensembl))
bound$`INPUT cpm`<-bound_k2bko[idx,]$`INPUT cpm`
bound$`BCOR WT cpm`<-bound_k2bko[idx,]$`BCOR WT cpm`
bound$`BCOR K2BKO cpm`<-bound_k2bko[idx,]$`BCOR K2BKO cpm`

colnames(bound)<-gsub(" ","_",colnames(bound)) %>% gsub("\\+","w_RNF2",.) %>% gsub("\\-","wo_RNF2",.) %>%  gsub("_cpm","",.)
bound[grep("EOMES",bound$hgnc),]
write.csv(bound,file="BCOR_KDM2B_INPUT_cpms_for_2kp_strong_weak_TSS_091517.csv",quote=F,row.names=F)



bound %>% as.data.frame() %>% 
  dplyr::select(-hgnc,-ensembl) %>% 
  dplyr::select(INPUT_w_RNF2,INPUT_wo_RNF2,BCOR_w_RNF2,BCOR_wo_RNF2,BCOR_WT,BCOR_K2BKO,INPUT,class) %>% 
  tidyr::gather(chip,cpm,-class) %>%
  mutate(chip=factor(chip,levels=c("BCOR_wo_RNF2","INPUT_wo_RNF2","BCOR_w_RNF2","INPUT_w_RNF2","BCOR_K2BKO","BCOR_WT","INPUT"))) %>% 
  mutate(cpm=log2(cpm)) %>%
  ggplot(aes(x=cpm,y=chip,fill=chip)) +geom_density_ridges(scale=1.0) + theme_bw() + scale_fill_manual(values=cbPalette[c(4,2,7,2,5,6,2)]) + facet_grid(class ~ .) +
  xlab("Log2(cpm)") +ylab("") + ggtitle("BCOR ChIP Distributions in RNF2 and KDM2B Mutants") + 
  scale_y_discrete(expand = c(0.075, 0)) + 
  geom_density_ridges(rel_min_height = 0.1)

bound %>% as.data.frame() %>% 
  dplyr::select(-hgnc,-ensembl) %>% 
  dplyr::select(KDM2B_w_RNF2,KDM2B_wo_RNF2,class) %>% 
  tidyr::gather(chip,cpm,-class) %>%
  mutate(chip=factor(chip,levels=c("KDM2B_wo_RNF2","KDM2B_w_RNF2","BCOR_K2BKO","BCOR_WT"))) %>% 
  mutate(cpm=log2(cpm)) %>%
  ggplot(aes(x=cpm,y=chip,fill=chip)) + geom_density_ridges() + theme_bw() + 
  scale_fill_manual(values=cbPalette[c(4,7)]) + facet_grid(class ~ .) +
  xlab("Log2(cpm)") +ylab("") + ggtitle("KDM2B ChIP Distributions +/- RNF2")  
```

```{r figure4d_heatmaps,eval=T}

bound_strong_sort<-bound_strong %>% as.data.frame() %>% 
  dplyr::filter(class=="strong") %>%
  dplyr::select(hgnc,ensembl,BCOR_WT_cpm,BCOR_K2BKO_cpm,BCOR_WT_wide_cpm,BCOR_K2BKO_wide_cpm) %>% 
#  dplyr::mutate(e=log2(BCOR_K2BKO) - log2(INPUT)) %>% 
  dplyr::mutate(e=log2(BCOR_K2BKO_cpm) - log2(BCOR_WT_cpm)) %>% 
  dplyr::mutate(e_wide=log2(BCOR_K2BKO_wide_cpm) - log2(BCOR_WT_wide_cpm)) %>% 
  dplyr::arrange(-BCOR_K2BKO_wide_cpm) 
  #dplyr::select(ensembl,e,BCOR_K2BKO)


head(bound_strong_sort)

bound_strong_sort$kdm2b_ex<-res_kdm2b[bound_strong_sort$ensembl,"log2FoldChange_k2bKO"]
bound_strong_sort$rnf2r_ex<-res_rabsh[bound_strong_sort$ensembl,"log2FoldChange_rabsh"]
bound_strong_sort$bcor_ex<-res_bcorsh[bound_strong_sort$ensembl,"log2FoldChange_bcorsh"]

#Match classification status
summary(idx<-match(bound_strong_sort$ensembl,tss_classified$ensembl))
bound_strong_sort$deg_bcor<-tss_classified[idx]$deg_bcor
bound_strong_sort$deg_k2bKO<-tss_classified[idx]$deg_k2bKO
bound_strong_sort$deg_rab<-tss_classified[idx]$deg_rab
table(bound_strong_sort$deg_bcor)
table(bound_strong_sort$deg_k2bKO)
table(bound_strong_sort$deg_rab)


bound_strong_sort[grep("ENSG00000152977",bound_strong_sort$ensembl),] #ZIC1
bound_strong_sort[grep("ENSG00000163508",bound_strong_sort$ensembl),] # EOMES
bound_strong_sort[grep("ENSG00000156574",bound_strong_sort$ensembl),] # NODAL
#barplot(bound_strong_sort$e)

#plot(log2(bound_strong_sort$BCOR_WT),log2(bound_strong_sort$BCOR_K2BKO),cex =0.5,pch=16)
#plot(log2(bound_strong_sort$BCOR_WT),bound_strong_sort$e,cex =0.5,pch=16)
#cor(log2(bound_strong_sort$BCOR_WT),log2(bound_strong_sort$BCOR_w_RNF2),method="spearman")
plot(rep(1,10),cex=2,pch=16,col=colorRampPalette(brewer.pal(11,"PRGn"))(10))
plot(rep(1,10),cex=2,pch=16,col=colorRampPalette(brewer.pal(11,"RdBu"))(10))
plot(rep(1,3),cex=5,pch=16,col=c("#67001F","#B51F2E","#EF3A38")) #Reds
#colors <- colorRampPalette(c("#A2AED9","snow","#EF3A38"))(200)
#colors<-rev(colorRampPalette(brewer.pal(11,"PRGn"))(200))
colors <- colorRampPalette(c("#A2AED9","snow","#B51F2E"))(200)
expression_df<- bound_strong_sort[,c("bcor_ex","kdm2b_ex")]
rownames(expression_df)<-bound_strong_sort$ensembl
annotation_df<- bound_strong_sort[,c("deg_k2bko","deg_bcor")]
rownames(annotation_df)<-bound_strong_sort$ensembl
annotation_df$deg_bcor<-factor(annotation_df$deg_bcor,levels=c("UP","NR","DN","XX"))
annotation_df$deg_k2bko<-factor(annotation_df$deg_k2bko,levels=c("UP","NR","DN","XX"))

ann_colors = list(
    deg_bcor= c(UP = "#EF3A38", NR = "white",DN="#A2AED9",XX="white"),
    deg_k2bko = c(UP = "#EF3A38", NR = "white",DN="#A2AED9",XX="white")
)

#use breaklist to set midpoint to 0
myBreaksList <- c(seq(min(expression_df,na.rm=T), 0, length.out=ceiling(length(colors)/2) + 1), 
              seq(max(expression_df,na.rm=T)/length(colors), max(expression_df,na.rm=T), length.out=floor(length(colors)/2)))

pheatmap(expression_df,cluster_rows=F,cluster_cols=F,labels_row="",col=colors,
         breaks=myBreaksList)

png(paste0("Figure4d_expression_heatmap_",ts,".png"),width=1,height=8,units="in",res=300)
pheatmap(expression_df,cluster_rows=F,cluster_cols=F,labels_row="",col=colors,breaks=myBreaksList)
dev.off()

boxplot(bound_strong_sort[,c("BCOR_WT","BCOR_K2BKO")],cex=0.5,pch=16)

colors2=colorRampPalette(brewer.pal(11,"PRGn"))(200)
#colors2 <- colorRampPalette(c("black","snow","#0052bf"))(200)

chip_df<- bound_strong_sort[,c("BCOR_WT_cpm","BCOR_K2BKO_cpm","BCOR_WT_wide_cpm","BCOR_K2BKO_wide_cpm","e","e_wide")] %>% 
    mutate(BCOR_K2BKO=log2(BCOR_K2BKO_cpm)) %>% 
    mutate(BCOR_WT=log2(BCOR_WT_cpm)) %>% 
    mutate(BCOR_K2BKO_wide=log2(BCOR_K2BKO_wide_cpm)) %>% 
    mutate(BCOR_WT_wide=log2(BCOR_WT_wide_cpm)) %>% 
    select(BCOR_WT,BCOR_K2BKO,BCOR_WT_wide,BCOR_K2BKO_wide,e,e_wide)

chip_df<- bound_strong_sort[,"e",drop=F]
rownames(chip_df)<-bound_strong_sort$ensembl

#use breaklist to set midpoint to 0
myBreaksList2 <- c(seq(min(chip_df,na.rm=T), 0, length.out=ceiling(length(colors2)/2) + 1), 
              seq(max(chip_df,na.rm=T)/length(colors2), max(chip_df,na.rm=T), length.out=floor(length(colors2)/2)))

png(paste0("Figure4d_chip_heatmap_",ts,".png"),width=1,height=8,units="in",res=300)
pheatmap(chip_df,cluster_rows=F,cluster_cols=F,labels_row="",col=colors2,breaks=myBreaksList2)
dev.off()


pheatmap(chip_df[,c(1,2,5)],cluster_rows=F,cluster_cols=F,labels_row="",col=colors2,breaks=myBreaksList2,annotation_row=annotation_df,annotation_colors = ann_colors)

png("Figure4d_chip_heatmap.png",width=1,height=8,units="in",res=300)
pheatmap(chip_df,cluster_rows=F,cluster_cols=F,labels_row="",col=colors2,breaks=myBreaksList2)
dev.off()


#(k2bko_fileset<-countFileset(new("fileset", filename=c("../hESC/H1_wt_BCOR_P36_Index15_S14.fastq.bam",
#"../hESC/H1_k2bko_BCOR_P36_Index17_S16.fastq.bam"),
#           labels=c("BCOR WT","BCOR K2BKO"))))

summary(idx<-match(bound_strong_sort$ensembl,tss_classified$ensembl))
#system.time(k2bko<-tornado(tss_classified[idx],pad=9000,k2bko_fileset,ord=0,color="#0052bf",window=100)) #blue
system.time(k2bko<-tornado(tss_classified[idx],pad=1000,k2bko_fileset,ord=0,color="#00441B",window=100)) #green
ggsave(file=paste0("figure4d_k2bko_bound_",ts,".png"),plot=k2bko+theme(panel.border = element_blank()),width=5,height=8)

#plot "e" by itself in heatmap

```

```{r}
bound$rfc<-log2(bound$BCOR_wo_RNF2)-log2(bound$BCOR_w_RNF2)
bound$kfc<-log2(bound$BCOR_K2BKO)-log2(bound$BCOR_WT)
cor(bound$rfc[bound$class=="strong"],bound$kfc[bound$class=="strong"],method="spearman")
cor(bound$rfc[bound$class=="weak"],bound$kfc[bound$class=="weak"],method="spearman")
cor(bound$rfc,bound$kfc,method="spearman")
bound[grep("ONECUT1",bound$hgnc),]
bound[grep("NODAL",bound$hgnc),]
bound[grep("EOMES",bound$hgnc),]

plot(bound$rfc[bound$class=="strong"],bound$kfc[bound$class=="strong"],cex=0.5,pch=16)
plot(bound$rfc[bound$class=="weak"],bound$kfc[bound$class=="weak"],cex=0.5,pch=16)

bound<-as.data.frame(bound)
#add deg from expression to color scatterplots
summary(idx<-match(bound$ensembl,tss_classified$ensembl))
bound$deg<-tss_classified[idx,]$deg

bound[bound$class=="strong",] %>% as.data.frame() %>% 
  dplyr::select(rfc,kfc,deg) %>% 
  dplyr::filter(deg=="UP" | deg=="DN") %>% 
  ggplot(aes(x=rfc,y=kfc,colour=deg)) + scale_colour_manual(values=cbPalette[c(1,3,5,6)]) +
  geom_point() +theme_bw()



bound[bound$class=="strong",] %>% as.data.frame() %>% 
  dplyr::select(rfc,kfc) %>% 
  arrange(rfc-kfc) %>% 
  t %>% 
  pheatmap(cluster_rows=F,cluster_cols=F,labels_col="",labels_row=c("BCOR ChIP FC in RNF2-R","BCOR ChIP FC in KDM2B KO"),col=rev(colorRampPalette(brewer.pal(11,"RdBu"))(200)))

#all.equal(rownames(df),rownames(temp))

ann_colors = list(deg = c("white", "firebrick","green","red"))
pheatmap(temp[,1:2],cluster_rows=F,cluster_cols=F,labels_row="",annotation_row = df,annotation_colors = ann_colors)

temp<-bound[bound$class=="strong",]
temp$kdm2b_ex<-res_kdm2b[temp$ensembl,"log2FoldChange_k2bKO"]
temp$rnf2r_ex<-res_rabsh[temp$ensembl,"log2FoldChange_rabsh"]
temp$bcor_ex<-res_bcorsh[temp$ensembl,"log2FoldChange_bcorsh"]
head(temp)
temp[grep("ONECUT1",temp$hgnc),]

temp<-temp %>% as.data.frame() %>% 
  dplyr::select(rfc,kfc,kdm2b_ex,rnf2r_ex,bcor_ex,hgnc) %>% 
  arrange(rfc-kfc) 
barplot(temp$rfc,main="rfc")
barplot(temp$rnf2r_ex,main="rnf2-r",ylim=c(-8,8))
barplot(temp$kfc,main="kfc")
barplot(temp$kdm2b_ex,main="kdm2b_ko",ylim=c(-8,8))
barplot(temp$bcor_ex,main="bcor_sh",ylim=c(-8,8))
sum(is.na(temp$rnf2r_ex))
temp<-temp[!is.na(temp$kdm2b_ex),]


  temp[,c(1:2)] %>% t %>% 
  pheatmap(cluster_rows=F,cluster_cols=F,labels_col="",labels_row=c("BCOR ChIP FC in RNF2-R","BCOR ChIP FC in KDM2B KO"),col=rev(colorRampPalette(brewer.pal(11,"RdBu"))(200)))
  
  temp[,c(5,4,3)]  %>%  t %>% 
  pheatmap(cluster_rows=F,cluster_cols=F,labels_col="",labels_row=c("Expression FC in shBCOR","Expression FC in shRAB","Expression FC in KDM2BKO"),col=colorRampPalette(brewer.pal(11,"PRGn"))(200))
  
 

```


#metaplots
```{r}

temp<-tss_classified[tss_classified$class=="strong"]
temp2<-sample(tss_classified[tss_classified$class=="nobinding"],length(temp))

temp<-sample(tss_classified[tss_classified$class=="strong"],100)
temp2<-sample(tss_classified[tss_classified$class=="nobinding"],101)

gr<-GRangesList(list(temp,temp2))
names(gr)<-c("strong","nobinding")

twister(gr,dataset=k2b_v_bcor_fileset,pad=9000,color=c("cyan","purple"))
tornado(temp,dataset=bcor_fileset,pad=9000,ord=0,color="red2")

```

```{r}
summary(degs_bcor_up<-rownames(subset(res_bcorsh,padj_bcorsh < alpha & log2FoldChange_bcorsh > lfc_thres)))
summary(degs_bcor_nc<-rownames(subset(res_bcorsh,padj_bcorsh > alpha & abs(log2FoldChange_bcorsh) <= lfc_thres)))
summary(degs_bcor_dn<-rownames(subset(res_bcorsh,padj_bcorsh < alpha & log2FoldChange_bcorsh < -1*lfc_thres)))

summary(degs_k2bKO_up<-rownames(subset(res_kdm2b,padj_k2bKO < alpha & log2FoldChange_k2bKO > lfc_thres)))
summary(degs_k2bKO_nc<-rownames(subset(res_kdm2b,padj_k2bKO > alpha & abs(log2FoldChange_k2bKO) <= lfc_thres)))
summary(degs_k2bKO_dn<-rownames(subset(res_kdm2b,padj_k2bKO < alpha & log2FoldChange_k2bKO < -1*lfc_thres)))

summary(degs_rab_up<-rownames(subset(res_rabsh,padj_rabsh < alpha & log2FoldChange_rabsh > lfc_thres)))
summary(degs_rab_nc<-rownames(subset(res_rabsh,padj_rabsh > alpha & abs(log2FoldChange_rabsh) <= lfc_thres)))
summary(degs_rab_dn<-rownames(subset(res_rabsh,padj_rabsh < alpha & log2FoldChange_rabsh < -1*lfc_thres)))

#strong
length(temp_up<-tss_classified[tss_classified$class=="strong" & tss_classified$BCOR_fpkm > 2.5 & tss_classified$ensembl %in% degs_bcor_up])
length(temp_dn<-tss_classified[tss_classified$class=="strong" & tss_classified$H1P_fpkm > 2.5 & tss_classified$ensembl %in% degs_bcor_dn])
length(temp_nr<-tss_classified[tss_classified$class=="strong" & !(tss_classified$ensembl %in% degs_bcor_up | tss_classified$ensembl %in% degs_bcor_dn)])

gr_strong<-GRangesList(list(temp_up,temp_nr,temp_dn))
names(gr_strong)<-c("UP","NR","DN")
lapply(gr_strong,length)

#weak
length(temp_up<-tss_classified[tss_classified$class=="weak" & tss_classified$BCOR_fpkm > 2.5 & tss_classified$ensembl %in% degs_bcor_up])
length(temp_dn<-tss_classified[tss_classified$class=="weak" & tss_classified$H1P_fpkm > 2.5 & tss_classified$ensembl %in% degs_bcor_dn])
length(temp_nr<-tss_classified[tss_classified$class=="weak" & !(tss_classified$ensembl %in% degs_bcor_up | tss_classified$ensembl %in% degs_bcor_dn)])

gr_weak<-GRangesList(list(temp_up,temp_nr,temp_dn))
names(gr_weak)<-c("UP","NR","DN")
lapply(gr_weak,length)
names(gr_strong)<-paste0("s",names(gr_strong))
names(gr_weak)<-paste0("w",names(gr_weak))
x<-c(gr_strong,gr_weak)
lapply(x,length)
save(gr_strong,gr_weak,file="meta_grannges.rdata")

gg_pol2<-twister(gr,dataset=pol2_fileset,pad=9000)
ggsave(gg_pol2,file="gg_pol2.pdf",height=5,width=5)

#alternatively add "UP","DN","NR" classification to tss_classified
tss_classified$deg_bcor<-ifelse(tss_classified$BCOR_fpkm > 2.5 & tss_classified$ensembl %in% degs_bcor_up,"UP",
                           ifelse(tss_classified$H1P_fpkm > 2.5 & tss_classified$ensembl %in% degs_bcor_dn,"DN",
                           ifelse(!(tss_classified$ensembl %in% degs_bcor_up | tss_classified$ensembl %in% degs_bcor_dn),"NR","XX")))
table(tss_classified$deg_bcor)
table(tss_classified[tss_classified$class=="strong"]$deg_bcor)

tss_classified$deg_k2bKO<-ifelse(tss_classified$KDM2B_fpkm > 2.5 & tss_classified$ensembl %in% degs_k2bKO_up,"UP",
                           ifelse(tss_classified$WT_fpkm > 2.5 & tss_classified$ensembl %in% degs_k2bKO_dn,"DN",
                           ifelse(!(tss_classified$ensembl %in% degs_k2bKO_up | tss_classified$ensembl %in% degs_k2bKO_dn),"NR","XX")))
table(tss_classified$deg_k2bKO)
table(tss_classified[tss_classified$class=="strong"]$deg_k2bKO)

tss_classified$deg_rab<-ifelse(tss_classified$H1_RAB_fpkm > 2.5 & tss_classified$ensembl %in% degs_rab_up,"UP",
                           ifelse(tss_classified$H1_H1P_fpkm > 2.5 & tss_classified$ensembl %in% degs_rab_dn,"DN",
                           ifelse(!(tss_classified$ensembl %in% degs_rab_up | tss_classified$ensembl %in% degs_rab_dn),"NR","XX")))
table(tss_classified$deg_rab)
table(tss_classified[tss_classified$class=="strong"]$deg_rab)

boxplot(log2(tss_classified[tss_classified$class=="strong"]$BCOR_wide +1) ~ tss_classified[tss_classified$class=="strong"]$deg_bcor, 
        main="BCOR_wide",ylab="log2 (counts +1)")



boxplot(log2(tss_classified[tss_classified$class=="strong"]$H3K27me3_wide +1) ~ tss_classified[tss_classified$class=="strong"]$deg_bcor, 
        main="H3K27me3_wide",ylab="log2 (counts +1)")
boxplot(log2(tss_classified[tss_classified$class=="strong"]$H3K36me2_wide+1) ~ tss_classified[tss_classified$class=="strong"]$deg, 
        main="H3K36me2_wide",ylab="log2 (counts +1)")
boxplot(log2(tss_classified[tss_classified$class=="strong"]$CBX2_wide+1) ~ tss_classified[tss_classified$class=="strong"]$deg, 
        main="CBX2_wide",ylab="log2 (counts +1)")

sum(tss_classified$class=="strong" & tss_classified$deg=="UP")
mean(tss_classified[tss_classified$class=="strong" & tss_classified$deg=="UP"]$H3K27me3_wide)
sum(tss_classified$class=="strong" & tss_classified$deg=="DN")
mean(tss_classified[tss_classified$class=="strong" & tss_classified$deg=="DN"]$H3K27me3_wide)
sum(tss_classified$class=="strong" & tss_classified$deg=="NR")
mean(tss_classified[tss_classified$class=="strong" & tss_classified$deg=="NR"]$H3K27me3_wide)

twister(gr_strong,dataset=histones_fileset,pad=0,window=25)

table(tss_classified[tss_classified$class=="strong"]$deg)
table(tss_classified[tss_classified$class=="weak"]$deg)


twister(gr,dataset=rna_fileset,pad=9000)
twister(gr,dataset=input_fileset,pad=9000)
twister(gr,dataset=bcor_fileset,pad=9000)
twister(gr,dataset=ser7p_fileset,pad=9000,sample_name = "Ser7p")

twister(gr,dataset=pol2_fileset,pad=9000)



twister(gr,dataset=rna_fileset,pad=9000)
twister(gr,dataset=bcor_fileset,pad=9000)



twister(gr[1:2],dataset=rna_fileset,pad=9000)

twister(gr,dataset=bcor_fileset,pad=9000)

twister(gr[1:2],dataset=pol2_fileset,pad=9000)

twister(gr,dataset=cbx2rybp_fileset,pad=9000)
twister(gr,dataset=histones_fileset,pad=9000)
twister(gr,dataset=dnase_fileset,pad=9000)
twister(gr,dataset=cgi_fileset,pad=9000)

```

#scratch
```{r }
boxplot(log2(tss_classified[tss_classified$class=="strong"]$BCOR) ~ tss_classified[tss_classified$class=="strong"]$deg_bcor, 
        main="BCOR",ylab="log2 (counts +1)")

mean(x1<-log2(tss_classified[tss_classified$class=="strong" & tss_classified$deg_bcor=="UP" ]BCOR_wide))
mean(x2<-log2(tss_classified[tss_classified$class=="strong" & tss_classified$deg_bcor=="NR" ]$BCOR_wide))
mean(y1<-log2(tss_classified[tss_classified$class=="weak" & tss_classified$deg_bcor=="UP" ]$BCOR_wide))
mean(y2<-log2(tss_classified[tss_classified$class=="weak" & tss_classified$deg_bcor=="NR" ]$BCOR_wide))
ks.test(x1,x2)$p.value
ks.test(y1,y2)$p.value
#xyz<-list(Ser5p_UP=x1,Ser5p_DN=x2,Ser7p_UP=y1,Ser7p_DN=y2)
xyz<-list(BCOR_UP=x1,BCOR_NR=x2,wBCOR_UP=y1,wBCOR_NR=y2)
lapply(xyz,length)
boxplot(xyz)


length(y<-log2(tss_classified[tss_classified$class=="strong" & tss_classified$deg_bcor=="NR" ]$CBX2))
length(z<-log2(tss_classified[tss_classified$class=="strong" & tss_classified$deg_bcor=="UP" ]$CBX2))
length(w<-log2(tss_classified[tss_classified$class=="strong" & tss_classified$deg_bcor=="XX" ]$CBX2))

xyz<-list(DN=x,NR=y,UP=z,XX=w)
lapply(xyz,length)
boxplot(xyz)
-1*log10(suppressWarnings(ks.test(y,z)$p.value))
y<-y[is.finite(y)]
z<-z[is.finite(z)]
-1*log10(suppressWarnings(ks.test(y,z)$p.value))

-1*log10(t.test(y,z)$p.value)


length(x<-log2(tss_classified[tss_classified$class=="weak" & tss_classified$deg_bcor=="UP" ]$BCOR_wide))
length(y<-log2(tss_classified[tss_classified$class=="weak" & tss_classified$deg_bcor=="NR" ]$BCOR_wide))

suppressWarnings(ks.test(x,y,exact=T)$p.value)
ks.test(x,y)$p.value
t.test(x,y,exact=T)$p.value
hist(y,xlim=c(6,10),breaks=100)
plot(ecdf(y), xlim = range(c(y, z)),cex=0.5,col="red")
plot(ecdf(x), add = TRUE, lty = "dashed",cex=0.4,col="blue")

x<-rnorm(1000,0,1)
y<-rnorm(1000,0,1.5)
rafalib::shist(y, xlim = c(5,15),col="red")
rafalib::shist(x,col="blue",add=T)
-1*log10(ks.test(x,y)$p.value)

plot(ecdf(rnorm(1000)), xlim = c(-10,10),cex=0.5,col="red")
plot(ecdf(rnorm(100,0,1.5)), add = TRUE,cex=0.5,col="blue")
-1*log10(ks.test(rnorm(1000),rnorm(1000,0,1.3))$p.value)


temp<-mcols(tss_classified[tss_classified$class=="strong"])[c(25:40,48)]

#skip normalization by depth because comparisons are within 1 ChIP.
#for (i in seq_along(wildtype_fileset@filename)) {
#  print(colnum<-which(gsub("_wide","",colnames(temp))==wildtype_fileset@labels[i]))
# # colnames(temp)[colnum]<-paste0(k2bko_fileset@labels[i]," wide cpm")
#  temp[,colnum]<-round(1e6*temp[,colnum]/as.numeric(wildtype_fileset@count[i]),3)
#}
head(temp)

temp %>% as.data.frame() %>% 
  dplyr::filter(deg_bcor!="XX") %>% 
  tidyr::gather(chip,counts,-deg_bcor) %>% 
  dplyr::mutate(counts=log2(counts+1)) %>% 
  dplyr::mutate(chip=factor(gsub("_wide","",chip),levels=wildtype_fileset@labels )) %>% 
  dplyr::mutate(deg_bcor=factor(deg_bcor,levels=c("NR","DN","UP"))) %>% 
  ggplot(aes(y=counts,x=deg_bcor)) +geom_boxplot(aes(fill=deg_bcor)) +facet_grid(chip ~ .) + 
  scale_fill_manual(values=rev(c("#EF3A38","#BFBFBF","#A2AED9")),name="") + ggtitle("strong") +
  coord_flip() + theme_bw() + theme(strip.background = element_blank(),
                                    panel.border = element_blank(),
                         panel.grid.major=element_blank(),
                         panel.grid.minor=element_blank())
ggsave("strong_boxplots.pdf",height=8,width=4)
```


#Figure S2G - Metaplot Stats
```{r Figure_S2G}

ks_chip_test<-function(cls="strong",cat1="UP",cat2="NR",signal="BCOR_wide") {
#x1<-mcols(tss_classified) %>% as.data.frame() %>% 
#  dplyr::filter(class==cls) %>% 
#  dplyr::filter(deg_bcor==cat1) %>% 
#  dplyr::select(signal) %>% 
#x2<-mcols(tss_classified) %>% as.data.frame() %>% 
#  dplyr::filter(class==cls) %>% 
#  dplyr::filter(deg_bcor==cat2) %>% 
#  dplyr::select(signal) %>% 
#  dplyr::filter((signal))

x1<-log2(mcols(tss_classified)[tss_classified$class==cls & tss_classified$deg_bcor==cat1,signal])
x2<-log2(mcols(tss_classified)[tss_classified$class==cls & tss_classified$deg_bcor==cat2,signal])

x1<-x1[is.finite(x1)]
x2<-x2[is.finite(x2)]
     
return(suppressWarnings(ks.test(x1,x2,exact=T)$p.value))

#return(suppressWarnings(t.test(na.omit(log2(x1[,1]+1)),na.omit(log2(x2[,1]+1)))$p.value))
}

ks_chip_test(cls="weak",cat1="UP",cat2="NR",signal="BCOR_wide")
options("scipen"=-100, "digits"=4)
df<-matrix(nrow=18,ncol=3)
colnames(df)<-c("UP vs DN","UP vs NR","DN vs NR")
#rownames(df)<-colnames(mcols(tss_classified))[5:22]
rownames(df)<-colnames(mcols(tss_classified))[23:40]

for (j in rownames(df)) {
  for (i in colnames(df)) {
  c1=substr(i,1,2)
  c2=substr(i,7,8)
  #print(paste0("c1:  ",c1," c2:  ",c2," ks.test:"))
  #print(paste0("c1:  ",c1," c2:  ",c2," ks.test:",ks_chip_test(cat1=c1,cat2=c2)))
  df[j,i]<-ks_chip_test(cat1=c1,cat2=c2,signal=j,cls="strong")
}
}

options("scipen"=999, "digits"=4)
#pheatmap(df,cluster_rows = F,cluster_cols = F,main="Strong Wide KS test")
strong_wide_ks<-df
colnames(strong_wide_ks)<-paste0(colnames(strong_wide_ks)," : strong")

# now do weak
options("scipen"=-100, "digits"=4)
df<-matrix(nrow=18,ncol=3)
colnames(df)<-c("UP vs DN","UP vs NR","DN vs NR")
#rownames(df)<-colnames(mcols(tss_classified))[5:22]
rownames(df)<-colnames(mcols(tss_classified))[23:40]

for (j in rownames(df)) {
  for (i in colnames(df)) {
  c1=substr(i,1,2)
  c2=substr(i,7,8)
  #print(paste0("c1:  ",c1," c2:  ",c2," ks.test:"))
  #print(paste0("c1:  ",c1," c2:  ",c2," ks.test:",ks_chip_test(cat1=c1,cat2=c2)))
  df[j,i]<-ks_chip_test(cat1=c1,cat2=c2,signal=j,cls="weak")
}
}

options("scipen"=999, "digits"=4)
#pheatmap(df,cluster_rows = F,cluster_cols = F,main="Weak Wide KS test")
weak_wide_ks<-df
colnames(weak_wide_ks)<-paste0(colnames(weak_wide_ks)," : weak")

wide_ks<-cbind(strong_wide_ks,weak_wide_ks)
#pheatmap(wide_ks,cluster_rows = F,cluster_cols = F,main="Wide KS test")

#pheatmap(narrow_ks,cluster_rows = F,cluster_cols = F,main="Narrow KS test")
#pheatmap(narrow_t,cluster_rows = F,cluster_cols = F,main="Narrow t-test")
#pheatmap(wide_ks,cluster_rows = F,cluster_cols = F,main="Wide KS test")
#pheatmap(wide_t,cluster_rows = F,cluster_cols = F,main="Wide t-test")

colors <- colorRampPalette(c("white","#fe3233"))(3)
colors <- colorRampPalette(c("white","red"))(3)
plot(rep(1,3),cex=2,pch=16,col=colors)

temp<-wide_ks %>% as.data.frame %>% 
  tibble::rownames_to_column(var="chip") %>% 
  dplyr::mutate(chip=gsub("_wide","",chip)) %>% 
  dplyr::mutate(chip=factor(chip,levels=rev(gsub("_wide","",colnames(mcols(tss_classified))[23:40])))) %>% 
  tidyr::gather(cat,pval,-chip) %>%
  dplyr::mutate(cat=factor(cat,levels=c("UP vs NR : strong","UP vs DN : strong","DN vs NR : strong","UP vs NR : weak","UP vs DN : weak","DN vs NR : weak"))) %>% 
  #dplyr::filter(pval > 0) %>% 
  dplyr::mutate(pval=round(-1*log10(pval),3)) %>% 
  dplyr::mutate(pval=replace(pval,is.infinite(pval),NA))

  temp_plot<-ggplot(temp,aes(x=cat,y=chip)) + geom_tile(aes(fill=pval))+
  scale_fill_gradient2(low=colors[1],mid=colors[2], high=colors[3],midpoint = 10,  name = "-Log10(p-value)")+
  geom_tile(data = subset(temp,  is.na(pval)), aes(colour = "Infinity"),linetype = 0, fill = "darkred") +
  xlab("")+ylab("")+
    geom_text(aes(x=cat,y=chip,label=pval),size=2) +
        theme_bw() + theme(strip.background = element_blank(),
                           axis.text.x = element_text(angle = 90, hjust = 1),
                                          axis.ticks.y=element_blank(),
                                          panel.grid.major=element_blank(),
                                          panel.grid.minor=element_blank())

ggsave(temp_plot,file="Figure_S2G_kstest.pdf",height=5,width=6)

```


#correlation matrix
https://stackoverflow.com/questions/41339515/single-correlation-heatmap-in-r-with-two-different-correlation-matrix-but-same
```{r}
load("../hESC/tss_classified_081817.rdata")
head(tss_classified)
#mat_strong<-as.matrix(mcols(tss_classified[tss_classified$class=="strong"])[,c(7:22)])
mat_strong<-as.matrix(mcols(tss_classified[tss_classified$class=="strong"])[,c(25:29,31:40)])
cor_strong <- cor(mat_strong,method="spearman")
#mat_weak<-as.matrix(mcols(tss_classified[tss_classified$class=="weak"])[,c(7:22)])
mat_weak<-as.matrix(mcols(tss_classified[tss_classified$class=="weak"])[,c(25:29,31:40)])
cor_weak <- cor(mat_weak,method="spearman")



cor_combined<-cor_weak
cor_combined[upper.tri(cor_combined)] <- cor_strong[upper.tri(cor_strong)]
diag(cor_combined) <- 0
#corrplot(cor_combined, method="circle")
mybreaksList = seq(-1, 1, by = 1/200)

cor_strong %>% 
pheatmap(cluster_rows=T,cluster_cols=T,col=rev(colorRampPalette(brewer.pal(11,"RdBu"))(length(mybreaksList))),breaks=mybreaksList)

cor_weak %>% 
pheatmap(cluster_rows=T,cluster_cols=T,col=rev(colorRampPalette(brewer.pal(11,"RdBu"))(length(mybreaksList))),breaks=mybreaksList)


```

```{r}
gr<-GRanges(seqnames="20",IRanges(start=22522676,end=22522863),strand="*")
seqlevelsStyle(gr)<-"UCSC"
twister(gr,dataset=histones_fileset,pad=0,window=1)

sbp_temp<-ScanBamParam(what = c("pos", "rname", "strand","qwidth"), which = gr,
                          flag = scanBamFlag(isUnmappedQuery = FALSE))

pileup("../hESC/H1_H3K27me3_ENCFF171RVK.bam" , scanBamParam=sbp_temp, pileupParam=p_param)

(h3k27me3_fileset<-countFileset(new("fileset", filename=c( "../hESC/H1_H3K27me3_ENCFF171RVK.bam"),labels=c("H3K27me3"))))
twister(gr_strong,dataset=h3k27me3_fileset,pad=0,window=1)
blah2<-twister(gr_strong,dataset=h3k27me3_fileset,pad=9000,window=25)
blah2 %>% group_by(genotype,whichgr) %>% summarize(sum=sum(count)/36) %>% ungroup

(h3k36me2_fileset<-countFileset(new("fileset",filename=c("../hESC/H3K36me2_ChIP_zheng_GCCAAT_bothruns.R1_trimmed.fastq.bam"),labels=c("H3K36me2"))))
twister(gr_strong,dataset=h3k36me2_fileset,pad=9000,window=25)

(cgi_fileset<-countFileset(new("fileset",filename=c("../hESC/cgi_hg38_canon_ncbi.bam"),labels=c("CGI"))))
twister(gr_strong,dataset=cgi_fileset,pad=9000,window=25)

twister(gr_strong,dataset=rna_fileset,pad=9000,window=25)
```

```{r}


files<-gsub("\\.bam","",sapply(strsplit(dataset@filename,"\\/"), function (x) x[length(x)]))

plot_twists<-function (i,env=parent.frame()) {
  temp_dataset<-new("fileset", filename=dataset@filename[which(files==i)],
                    labels=dataset@labels[which(files==i)],
                    count=dataset@count[which(files==i)])
  twister(gr_strong,dataset=temp_dataset,pad=9000,window=25)
}

my_plots<-lapply(files,plot_twists)
pdf("test_myplots.pdf",height=6,width=6)
my_plots
dev.off()

```

```{r}
bound %>% as.data.frame() %>% 
  dplyr::select(-hgnc,-ensembl) %>% 
    dplyr::filter(class=="strong") %>% 
  dplyr::select(INPUT_w_RNF2,BCOR_WT,BCOR_K2BKO) %>% 
  tidyr::gather(chip,cpm) %>%
  mutate(chip=factor(chip,levels=c("BCOR_K2BKO","BCOR_WT","INPUT_w_RNF2"))) %>% 
  mutate(cpm=log2(cpm)) %>%
  ggplot(aes(x=cpm,y=chip,fill=chip)) +geom_density_ridges(scale=1.0) + theme_bw() +   scale_fill_manual(values=cbPalette[c(4,7,2)]) + 
  scale_y_discrete(expand = c(0.075, 0)) +theme


bound %>% as.data.frame() %>% 
  dplyr::select(-hgnc,-ensembl) %>% 
    dplyr::filter(class=="strong") %>% 
  dplyr::select(INPUT_w_RNF2,BCOR_w_RNF2,BCOR_wo_RNF2) %>% 
  tidyr::gather(chip,cpm) %>%
  mutate(chip=factor(chip,levels=c("BCOR_wo_RNF2","BCOR_w_RNF2","INPUT_w_RNF2"))) %>% 
  mutate(cpm=log2(cpm)) %>%
  ggplot(aes(x=cpm,y=chip,fill=chip)) +geom_density_ridges(scale=1.0) + theme_bw() +   scale_fill_manual(values=cbPalette[c(4,7,2)]) + 
  scale_y_discrete(expand = c(0.075, 0)) 
```

GO Analysis for RNA-Seq
```{r}
load("GO_data.rdata")

#subset GO terms to genes expressed shRNA
expressed_cutoff<-10
(length(expressed_genes<-rownames(res[res$baseMean_bcorsh > expressed_cutoff | res$baseMean_rabsh > expressed_cutoff | res$baseMean_k2bKO > expressed_cutoff,])))

dim(goa_human<-goa_human[goa_human$ensembl %in% expressed_genes,])

#reduce to term and gene
dim(goa_human<-goa_human[!is.na(goa_human$ensembl),c("ensembl","GO ID")])

#remove duplicated entries (possibly from multiple lines of evidence)
dim(goa_human<-goa_human[!duplicated(goa_human),])

#shorted expressed genes to just those that we have ontology information for.
length(expressed_genes<-expressed_genes[expressed_genes %in% unique(goa_human$ensembl)])

#make a list
goa_human.list<-split(goa_human$`GO ID`,goa_human$ensembl)

#test
#go_terms[goa_human.list[[temp[grep("FOXA2",temp$SYMBOL),"GENEID"]]]]
#go_terms[goa_human.list[[temp[grep("^BCOR$",temp$SYMBOL),"GENEID"]]]]
#go_terms[goa_human.list[[temp[grep("^DMRT1$",temp$SYMBOL),"GENEID"]]]]

#goa_human[grep("GO:0050911",goa_human$`GO ID`),]$ensembl
#AnnotationDbi::select(ens88,keys=unique(goa_human[grep("GO:0050911",goa_human$`GO ID`),]$ensembl),keytype="GENEID",columns=c("SYMBOL"))

#Get bias data
#bd<-sum(width(reduce(exonsBy(ens88,by="gene"))))
#assays(dds)$avgTxLength["ENSG00000183337",]

bd<-rowMeans(assays(dds)$avgTxLength)
bd[temp[grep("^DMRT1$",temp$SYMBOL),"GENEID"]]
bd<-bd[expressed_genes]


#run goseq on 3 categories
sum(degs.bcor_up<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$deg_bcor=="UP"]))
names(degs.bcor_up)<-expressed_genes
table(degs.bcor_up)

sum(degs.bcor_dn<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$deg_bcor=="DN"]))
names(degs.bcor_dn)<-expressed_genes
table(degs.bcor_dn)

sum(degs.bcor_nr<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$deg_bcor=="NR"]))
names(degs.bcor_nr)<-expressed_genes
table(degs.bcor_nr)

pwf.bcor_up<-nullp(degs.bcor_up,bias.data=bd)
pwf.bcor_dn<-nullp(degs.bcor_dn,bias.data=bd)
pwf.bcor_nr<-nullp(degs.bcor_nr,bias.data=bd)

GO.wall.bcor_up.goa<-goseq(pwf.bcor_up,gene2cat=goa_human.list)
GO.wall.bcor_dn.goa<-goseq(pwf.bcor_dn,gene2cat=goa_human.list)
GO.wall.bcor_nr.goa<-goseq(pwf.bcor_nr,gene2cat=goa_human.list)
View(GO.wall.bcor_up.goa)
View(GO.wall.bcor_dn.goa)
View(GO.wall.bcor_nr.goa)


write.csv(GO.wall.bcor_up.goa[GO.wall.bcor_up.goa$over_represented_pvalue < 0.05,],file="GO_wall_bcor_up_goa.csv",quote=F,row.names = F)
write.csv(GO.wall.bcor_dn.goa[GO.wall.bcor_dn.goa$over_represented_pvalue < 0.05,],file="GO_wall_bcor_dn_goa.csv",quote=F,row.names = F)
write.csv(GO.wall.bcor_nr.goa[GO.wall.bcor_nr.goa$over_represented_pvalue < 0.05,],file="GO_wall_bcor_nr_goa.csv",quote=F,row.names = F)

revigo.bcor_up<-read.csv("revigo_bcor_up.csv",stringsAsFactors = F,header=T)
revigo.bcor_up$bcor_deg<-"UP"
revigo.bcor_dn<-read.csv("revigo_bcor_dn.csv",stringsAsFactors = F,header=T)
revigo.bcor_dn$bcor_deg<-"DN"
revigo.bcor_nr<-read.csv("revigo_bcor_nr.csv",stringsAsFactors = F,header=T)
revigo.bcor_nr$bcor_deg<-"NR"


bcor_deg_go<-rbind(revigo.bcor_up,revigo.bcor_dn) %>% 
  dplyr::select(description,log10.p.value,bcor_deg) %>% 
  dplyr::mutate(log10.p.value = -1* log10.p.value) %>% 
  dplyr::filter(log10.p.value > 4) %>% 
  dplyr::mutate(bcor_deg=factor(bcor_deg,levels=rev(c("UP","NR","DN")))) %>% 
  dplyr::arrange(bcor_deg,log10.p.value) %>% 
  dplyr::mutate(description=factor(description,levels=description)) %>% 
ggplot(aes(x=description,y=log10.p.value,fill=bcor_deg)) + 
  geom_bar(stat="identity") +
  scale_fill_manual(values=c("#A2AED9","#EF3A38"),guide=FALSE) +
  coord_flip() + ylab("-Log10( p-value)") +
  theme_bw() +
  theme(strip.background = element_blank(), 
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_blank())

ggsave(bcor_deg_go,file="bcor_deg_go_woNR.pdf",height=12,width=8)

```


#KDM2B GO terms
```{r}
sum(degs.k2bKO_up<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$deg_k2bKO=="UP"]))
names(degs.k2bKO_up)<-expressed_genes
table(degs.k2bKO_up)

sum(degs.k2bKO_dn<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$deg_k2bKO=="DN"]))
names(degs.k2bKO_dn)<-expressed_genes
table(degs.k2bKO_dn)

sum(degs.k2bKO_nr<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$deg_k2bKO=="NR"]))
names(degs.k2bKO_nr)<-expressed_genes
table(degs.k2bKO_nr)

pwf.k2bKO_up<-nullp(degs.k2bKO_up,bias.data=bd)
pwf.k2bKO_dn<-nullp(degs.k2bKO_dn,bias.data=bd)
pwf.k2bKO_nr<-nullp(degs.k2bKO_nr,bias.data=bd)

GO.wall.k2bKO_up.goa<-goseq(pwf.k2bKO_up,gene2cat=goa_human.list)
GO.wall.k2bKO_dn.goa<-goseq(pwf.k2bKO_dn,gene2cat=goa_human.list)
GO.wall.k2bKO_nr.goa<-goseq(pwf.k2bKO_nr,gene2cat=goa_human.list)
View(GO.wall.k2bKO_up.goa)
View(GO.wall.k2bKO_dn.goa)
View(GO.wall.k2bKO_nr.goa)


write.csv(GO.wall.k2bKO_up.goa[GO.wall.k2bKO_up.goa$over_represented_pvalue < 0.05,],file="GO_wall_k2bKO_up_goa.csv",quote=F,row.names = F)
write.csv(GO.wall.k2bKO_dn.goa[GO.wall.k2bKO_dn.goa$over_represented_pvalue < 0.05,],file="GO_wall_k2bKO_dn_goa.csv",quote=F,row.names = F)
write.csv(GO.wall.k2bKO_nr.goa[GO.wall.k2bKO_nr.goa$over_represented_pvalue < 0.05,],file="GO_wall_k2bKO_nr_goa.csv",quote=F,row.names = F)

revigo.k2bKO_up<-read.csv("revigo_kdm2b_up.csv",stringsAsFactors = F,header=T)
revigo.k2bKO_up$k2bKO_deg<-"UP"
revigo.k2bKO_dn<-read.csv("revigo_kdm2b_dn.csv",stringsAsFactors = F,header=T)
revigo.k2bKO_dn$k2bKO_deg<-"DN"
revigo.k2bKO_nr<-read.csv("revigo_kdm2b_nr.csv",stringsAsFactors = F,header=T)
revigo.k2bKO_nr$k2bKO_deg<-"NR"


k2bKO_deg_go<-rbind(revigo.k2bKO_up,revigo.k2bKO_dn) %>% 
  dplyr::select(description,log10.p.value,k2bKO_deg) %>% 
  dplyr::mutate(log10.p.value = -1* log10.p.value) %>% 
  dplyr::filter(log10.p.value > 4) %>% 
  dplyr::mutate(k2bKO_deg=factor(k2bKO_deg,levels=rev(c("UP","NR","DN")))) %>% 
  dplyr::arrange(k2bKO_deg,log10.p.value) %>% 
  dplyr::mutate(description=factor(description,levels=description)) %>% 
ggplot(aes(x=description,y=log10.p.value,fill=k2bKO_deg)) + 
  geom_bar(stat="identity") +
  scale_fill_manual(values=c("#A2AED9","#EF3A38"),guide=FALSE) +
  coord_flip() + ylab("-Log10( p-value)") +
  theme_bw() +
  theme(strip.background = element_blank(), 
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_blank())

ggsave(k2bKO_deg_go,file="k2bKO_deg_go_woNR.pdf",height=12,width=8)
```

#RAB GO terms
```{r}
sum(degs.rab_up<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$deg_rab=="UP"]))
names(degs.rab_up)<-expressed_genes
table(degs.rab_up)

sum(degs.rab_dn<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$deg_rab=="DN"]))
names(degs.rab_dn)<-expressed_genes
table(degs.rab_dn)

sum(degs.rab_nr<-as.numeric(expressed_genes %in% tss_classified$ensembl[tss_classified$deg_rab=="NR"]))
names(degs.rab_nr)<-expressed_genes
table(degs.rab_nr)

pwf.rab_up<-nullp(degs.rab_up,bias.data=bd)
pwf.rab_dn<-nullp(degs.rab_dn,bias.data=bd)
pwf.rab_nr<-nullp(degs.rab_nr,bias.data=bd)

GO.wall.rab_up.goa<-goseq(pwf.rab_up,gene2cat=goa_human.list)
GO.wall.rab_dn.goa<-goseq(pwf.rab_dn,gene2cat=goa_human.list)
GO.wall.rab_nr.goa<-goseq(pwf.rab_nr,gene2cat=goa_human.list)
#View(GO.wall.rab_up.goa)
#View(GO.wall.rab_dn.goa)
#View(GO.wall.rab_nr.goa)


write.csv(GO.wall.rab_up.goa[GO.wall.rab_up.goa$over_represented_pvalue < 0.05,],file="GO_wall_rab_up_goa.csv",quote=F,row.names = F)
write.csv(GO.wall.rab_dn.goa[GO.wall.rab_dn.goa$over_represented_pvalue < 0.05,],file="GO_wall_rab_dn_goa.csv",quote=F,row.names = F)
write.csv(GO.wall.rab_nr.goa[GO.wall.rab_nr.goa$over_represented_pvalue < 0.05,],file="GO_wall_rab_nr_goa.csv",quote=F,row.names = F)

revigo.rab_up<-read.csv("revigo_rab_up.csv",stringsAsFactors = F,header=T)
revigo.rab_up$rab_deg<-"UP"
revigo.rab_dn<-read.csv("revigo_rab_dn.csv",stringsAsFactors = F,header=T)
revigo.rab_dn$rab_deg<-"DN"
revigo.rab_nr<-read.csv("revigo_rab_nr.csv",stringsAsFactors = F,header=T)
revigo.rab_nr$rab_deg<-"NR"


rab_deg_go<-rbind(revigo.rab_up,revigo.rab_nr,revigo.rab_dn) %>% 
  dplyr::select(description,log10.p.value,rab_deg) %>% 
  dplyr::mutate(log10.p.value = -1* log10.p.value) %>% 
  dplyr::filter(log10.p.value > 4) %>% 
  dplyr::mutate(rab_deg=factor(rab_deg,levels=rev(c("UP","NR","DN")))) %>% 
  dplyr::arrange(rab_deg,log10.p.value) %>% 
  dplyr::mutate(description=factor(description,levels=description)) %>% 
ggplot(aes(x=description,y=log10.p.value,fill=rab_deg)) + 
  geom_bar(stat="identity") +
  scale_fill_manual(values=c("#A2AED9","#EF3A38"),guide=FALSE) +
  coord_flip() + ylab("-Log10( p-value)") +
  theme_bw() +
  theme(strip.background = element_blank(), 
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_blank())

ggsave(rab_deg_go,file="rab_deg_go.pdf",height=12,width=8)
```

```{r figure4d}
head(degs.k2bKO_up)
load("bound_w_counts.rdata")

#test boxplot
length(kd2b_only<-bound[bound$class=="strong" & bound$deg_k2bKO=="UP" & bound$deg_bcor!="UP"])
length(bcor_only<-bound[bound$class=="strong" & bound$deg_k2bKO!="UP" & bound$deg_bcor=="UP"])
length(k2b_bcor<-bound[bound$class=="strong" & bound$deg_k2bKO=="UP" & bound$deg_bcor=="UP"])
length(not_up<-bound[bound$class=="strong" & bound$deg_k2bKO!="UP" & bound$deg_bcor!="UP"])


xyz<-list(bcor_only_wt=log2(bcor_only$BCOR_WT_cpm),bcor_only_mt=log2(bcor_only$BCOR_K2BKO_cpm),
          k2b_and_bcor_wt=log2(k2b_bcor$BCOR_WT_cpm),k2b_and_bcor_mt=log2(k2b_bcor$BCOR_K2BKO_cpm),
          k2b_only_wt=log2(kd2b_only$BCOR_WT_cpm),k2b_only_mt=log2(kd2b_only$BCOR_K2BKO_cpm),
          not_up_wt=log2(not_up$BCOR_WT_cpm),not_up_mt=log2(not_up$BCOR_K2BKO_cpm))
lapply(xyz,length)
lapply(xyz,mean)
boxplot(xyz,range=0,main="")
xyz<-list(bcor_only_wt=log2(bcor_only$BCOR_WT_wide_cpm),bcor_only_mt=log2(bcor_only$BCOR_K2BKO_wide_cpm),
          k2b_and_bcor_wt=log2(k2b_bcor$BCOR_WT_wide_cpm),k2b_and_bcor_mt=log2(k2b_bcor$BCOR_K2BKO_wide_cpm),
          k2b_only_wt=log2(kd2b_only$BCOR_WT_wide_cpm),k2b_only_mt=log2(kd2b_only$BCOR_K2BKO_wide_cpm),
          not_up_wt=log2(not_up$BCOR_WT_wide_cpm),not_up_mt=log2(not_up$BCOR_K2BKO_wide_cpm))
lapply(xyz,length)
lapply(xyz,mean)
boxplot(xyz,range=0,main="")



bound$kd2b_v_bcor<-ifelse(bound$deg_k2bKO=="UP" & bound$deg_bcor!="UP","k2b_only",
                            ifelse(bound$deg_k2bKO!="UP" & bound$deg_bcor=="UP","bcor_only",
                            ifelse(bound$deg_k2bKO=="UP" & bound$deg_bcor=="UP","k2b_bcor","not_up")))
table(bound[bound$class=="strong"]$kd2b_v_bcor)
table(bound[bound$class=="weak"]$kd2b_v_bcor)

mcols(bound) %>% as.data.frame() %>%
 # dplyr::filter(class=="strong") %>% 
  dplyr::select(kd2b_v_bcor,class,BCOR_WT_wide_cpm,BCOR_K2BKO_wide_cpm,BCOR_PLUS_wide_cpm,BCOR_MINUS_wide_cpm,KDM2B_PLUS_wide_cpm,KDM2B_MINUS_wide_cpm) %>% 
  tidyr::gather(chip,cpm,-kd2b_v_bcor,-class) %>% 
  dplyr::mutate(cpm=log2(cpm)) %>%
  dplyr::mutate(chip=gsub("_wide_cpm","",chip)) %>% 
  dplyr::mutate(chip=factor(chip,levels=c("BCOR_WT","BCOR_K2BKO","BCOR_PLUS","BCOR_MINUS","KDM2B_PLUS","KDM2B_MINUS"))) %>% 
  dplyr::filter(is.finite(cpm)) %>% 
  ggplot(aes(x=chip,y=cpm)) + 
  geom_boxplot(coef=5) +
  facet_grid(class ~ kd2b_v_bcor) +
  xlab("") + ylab(" Log2 (cpm)") +
  theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1))
  
#option 2
mcols(bound) %>% as.data.frame() %>%
 # dplyr::filter(class=="strong") %>% 
  dplyr::select(kd2b_v_bcor,class,BCOR_PLUS_wide_cpm,BCOR_MINUS_wide_cpm,KDM2B_PLUS_wide_cpm,KDM2B_MINUS_wide_cpm) %>% 
  tidyr::gather(chip,cpm,-kd2b_v_bcor,-class) %>% 
  dplyr::mutate(cpm=log2(cpm)) %>%
  dplyr::mutate(chip=gsub("_wide_cpm","",chip)) %>% 
  dplyr::mutate(chip=factor(chip,levels=c("BCOR_WT","BCOR_K2BKO","BCOR_PLUS","BCOR_MINUS","KDM2B_PLUS","KDM2B_MINUS"))) %>% 
  dplyr::filter(is.finite(cpm)) %>% 
  ggplot(aes(x=chip,y=cpm,fill=chip)) + 
  geom_boxplot(outlier.size = 0.5) +
  facet_grid(kd2b_v_bcor~class) +
  xlab("") + ylab(" Log2 (cpm)") + scale_fill_manual(values=rep(c("gray","purple"),2)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1),
                     strip.background = element_blank(), 
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank())

#option 3
mcols(bound) %>% as.data.frame() %>%
 # dplyr::filter(class=="strong") %>% 
  dplyr::select(kd2b_v_bcor,class,KDM2B_PLUS_wide_cpm,KDM2B_MINUS_wide_cpm) %>% 
  tidyr::gather(chip,cpm,-kd2b_v_bcor,-class) %>% 
  dplyr::mutate(cpm=log2(cpm)) %>%
  dplyr::mutate(chip=gsub("_wide_cpm","",chip)) %>% 
  dplyr::mutate(chip=factor(chip,levels=c("BCOR_WT","BCOR_K2BKO","BCOR_PLUS","BCOR_MINUS","KDM2B_PLUS","KDM2B_MINUS"))) %>% 
  dplyr::filter(is.finite(cpm)) %>% 
  ggplot(aes(x=chip+class,y=cpm,fill=chip)) + 
  geom_boxplot(outlier.size = 0.5) +
  #facet_grid(chip ~ class+kd2b_v_bcor) +
  facet_wrap(c("kd2b_v_bcor", "class"), labeller = "label_both",nrow=1) +
  xlab("") + ylab(" Log2 (cpm)") + scale_fill_manual(values=rep(c("gray","purple"),2)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1),
                     strip.background = element_blank(), 
       #              panel.border = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank())

#option 4
mcols(bound) %>% as.data.frame() %>%
 # dplyr::filter(class=="strong") %>% 
  dplyr::select(kd2b_v_bcor,class,BCOR_PLUS_wide_cpm,BCOR_MINUS_wide_cpm,KDM2B_PLUS_wide_cpm,KDM2B_MINUS_wide_cpm) %>% 
  tidyr::gather(chip,cpm,-kd2b_v_bcor,-class) %>% 
 # dplyr::mutate(class2=paste0(class,":",kd2b_v_bcor)) %>% 
  dplyr::rowwise() %>% 
  dplyr::mutate(antibody=stringr::str_split(chip,"_",simplify = T)[1]) %>% 
  dplyr::mutate(dox=stringr::str_split(chip,"_",simplify = T)[2]) %>% 
  dplyr::ungroup() %>%
  dplyr::mutate(x=factor(paste0(class,dox),levels=c("strongPLUS","strongMINUS","weakPLUS","weakMINUS"))) %>% 
  dplyr::mutate(cpm=log2(cpm)) %>%
  dplyr::mutate(chip=gsub("_wide_cpm","",chip)) %>% 
 # dplyr::mutate(chip=factor(chip,levels=c("BCOR_WT","BCOR_K2BKO","BCOR_PLUS","BCOR_MINUS","KDM2B_PLUS","KDM2B_MINUS"))) %>% 
  dplyr::filter(is.finite(cpm)) %>% 
  ggplot(aes(x=x,y=cpm,fill=chip)) + 
  geom_boxplot(outlier.size = 0.5) +
  facet_grid(antibody ~ kd2b_v_bcor) +
 # facet_wrap(c("kd2b_v_bcor", "class"), labeller = "label_both",nrow=1) +
  xlab("") + ylab(" Log2 (cpm)") + scale_fill_manual(values=rep(c("purple","gray"),2)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1),
                     strip.background = element_blank(), 
       #              panel.border = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank())

ggsave("Figure6_boxplot.pdf",width=8,height=4)

gr_k2b_v_bcor<-GRangesList(list(kd2b_only,bcor_only))
names(gr_k2b_v_bcor)<-c("KDM2B_Only","BCOR_Only")
lapply(gr_k2b_v_bcor,length)
save(gr_k2b_v_bcor,file="gr_k2b_v_bcor.rdata")


(k2b_v_bcor_fileset<-countFileset(new("fileset",filename=c("../hESC/H1_wt_BCOR_P36_Index15_S14.fastq.bam",
                                                         "../hESC/H1_egs_KDM2B_P36_Index11_S6.fastq.bam",
                                                         "../hESC/H3K36me2_ChIP_zheng_GCCAAT_bothruns.R1_trimmed.fastq.bam"),
                                    labels=c("BCOR","KDM2B","H3K36me2"))))
(k36me2<-countFileset(new("fileset",filename=c("../hESC/Input_DNA_zheng_CAGATC_bothruns.R1_trimmed.fastq.bam",
                                               "../hESC/H3K36me2_ChIP_zheng_GCCAAT_bothruns.R1_trimmed.fastq.bam"),
                                    labels=c("Input","H3K36me2"))))

(h1psh<-countFileset(new("fileset",filename=c("../hESC/1shH1P_input_CTTGTA_L001_R1_all.R1_trimmed.fastq.bam",
                                               "../hESC/1shH1P_k36_ACAGTG_L001_R1_all.R1_trimmed.fastq.bam"),
                                    labels=c("1shH1P-Input","1shH1P-H3K36me2"))))

length(random<-sample(tss_classified[tss_classified$class=="strong"],100))
twister(random,dataset=k36me2,pad=9000,window=25,color=c("blue"))
twister_diff(gr_k2b_v_bcor,dataset=k36me2,pad=9000,window=25,color=c("blue"))
tornado_diff(random,dataset=k36me2,pad=9000,window=25,color=c("blue"))

temp<-twister(gr_k2b_v_bcor,dataset=k2b_v_bcor_fileset,pad=9000,window=25)
temp + theme(strip.background = element_blank())
```

 
```{r}

#dds_tcrab3<-dds[,colData(dds)$time<=3 & colData(dds)$sample== "RAB" ]
#design(dds_tcrab3)<- ~ 1 + bs(time)
#dds_tcrab3 <- DESeq(dds_tcrab3, test="LRT",reduced=~1 )
#res_tcrab3 <- results(dds_tcrab3)
#colnames(res_tcrab3)<-paste(colnames(res_tcrab3),".rab3")

dds_tcrab6<-dds[,colData(dds)$time<=6 & colData(dds)$sample== "RAB" ]
design(dds_tcrab6)<- ~ 1 + bs(time)
dds_tcrab6 <- DESeq(dds_tcrab6, test="LRT",reduced=~1 )
res_tcrab6 <- results(dds_tcrab6)
colnames(res_tcrab6)<-paste(colnames(res_tcrab6),".rab6")

#dds_tcbcor3<-dds[,colData(dds)$time<=3 & colData(dds)$sample== "BCOR" ]
#design(dds_tcbcor3)<- ~ 1 + bs(time)
#dds_tcbcor3 <- DESeq(dds_tcbcor3, test="LRT",reduced=~1 )
#res_tcbcor3 <- results(dds_tcbcor3)
#colnames(res_tcbcor3)<-paste(colnames(res_tcbcor3),".bcor3")

dds_tcbcor6<-dds[,colData(dds)$time<=6 & colData(dds)$sample== "BCOR" ]
design(dds_tcbcor6)<- ~ 1 + bs(time)
dds_tcbcor6 <- DESeq(dds_tcbcor6, test="LRT",reduced=~1 )
res_tcbcor6 <- results(dds_tcbcor6)
colnames(res_tcbcor6)<-paste(colnames(res_tcbcor6),".bcor6")

#res_tc<-cbind(res_tcrab3,res_tcrab6,res_tcbcor3,res_tcbcor6)
res_tc<-cbind(res_tcrab6,res_tcbcor6)
idx<-match(rownames(res_tc),hgnc$ensembl_gene_id)
res_tc$hgnc<-hgnc[idx,"hgnc_symbol"]

#dim(res_tc_dif<-subset(res_tc,(abs(log2FoldChange..rab3) > 1 & padj..rab3 < 0.05) |
#                          (abs(log2FoldChange..rab6) > 1 & padj..rab6 < 0.05) |
#                          (abs(log2FoldChange..bcor3) > 1 & padj..bcor3 < 0.05) |
#                          (abs(log2FoldChange..bcor6) > 1 & padj..bcor6 < 0.05) ))
dim(res_tc_dif<-subset(res_tc, (abs(log2FoldChange..rab6) > 1 & padj..rab6 < 0.05) |
                          ((abs(log2FoldChange..bcor6) > 1 & padj..bcor6 < 0.05) ))

length(bcor_genes<-rownames(subset(as.data.frame(res_tcbcor6), padj..bcor6 < 0.05 )))  # all genes that change over time
length(rab_genes<-rownames(subset(as.data.frame(res_tcrab6), padj..rab6 < 0.05 )))  # all genes that change over time
length(kdm2b_genes<-tss_classified$ensembl[tss_classified$deg_k2bKO!="NR" ])  #kdm2b degs

length(bcor_genes<-bcor_genes[bcor_genes %in% tss_classified$ensembl[tss_classified$class!="nobinding"]])
length(rab_genes<-rab_genes[rab_genes %in% tss_classified$ensembl[tss_classified$class!="nobinding"]])
length(kdm2b_genes<-kdm2b_genes[kdm2b_genes %in% tss_classified$ensembl[tss_classified$class!="nobinding"]])

#length(unique(c(bcor_genes,rab_genes,kdm2b_genes)))
dim(f_diff<-f[unique(c(bcor_genes,rab_genes,kdm2b_genes)),c(1:7,16:22,25,14,15)])
table(apply(f_diff,1,function(x) sum(x > 2.5)))
dim(f_diff<-f_diff[apply(f_diff,1,function(x) sum(x > 2.5)) > 0,])
f_diff<-log2(f_diff+0.1)
#get fold change over t=0
head(f_diff)
temp<-cbind(f_diff[,2:7]-f_diff[,1],f_diff[,9:14]-f_diff[,8],f_diff[,16:17]-f_diff[,15])

#fpkm filter for UP in Venn Diagram
temp<-as.data.frame(temp)
length(bcor_genes<-bcor_genes[bcor_genes %in% rownames(temp[temp$bcor_1 > 1 | temp$bcor_2 > 1 | temp$bcor_3 > 1 | temp$bcor_4 > 1 | temp$bcor_5 > 1 | temp$bcor_6 > 1,])])
length(rab_genes<-rab_genes[rab_genes %in% rownames(temp[temp$r1b_1 > 1 | temp$r1b_2 > 1 | temp$r1b_3 > 1 | temp$r1b_4 > 1 | temp$r1b_5 > 1 | temp$r1b_6 > 1,])])
length(kdm2b_genes<-kdm2b_genes[kdm2b_genes %in% rownames(temp[temp$kdm2b1 > 1 | temp$kdm2b2 > 1,])])
#sanity check
pheatmap(temp[rownames(temp) %in% bcor_genes,],cluster_cols=F,show_rownames = F,scale="none",col=colors)
pheatmap(temp[rownames(temp) %in% rab_genes,],cluster_cols=F,show_rownames = F,scale="none",col=colors)
pheatmap(temp[rownames(temp) %in% kdm2b_genes,],cluster_cols=F,show_rownames = F,scale="none",col=colors)

temp2<-temp[rownames(temp) %in% c(bcor_genes,rab_genes,kdm2b_genes),]
pheatmap(temp2,cluster_cols=F,show_rownames = F,scale="none",col=colors)

write.csv(temp2,file="Figure6_heatmap_table.csv",quote=F, col.names = T, row.names = T)

png(paste0("Figure6_timecourse_heatmap_",ts,".png"),width=3,height=4,units="in",res=150)
pheatmap(temp2,cluster_cols=F,show_rownames = F,scale="none",col=colors)
dev.off()


overrideTriple<-T
grid.newpage();
vennplot <- draw.triple.venn(length(bcor_genes),
                             length(rab_genes),
                             length(kdm2b_genes),
                             sum(bcor_genes %in% rab_genes), #n12
                             sum(rab_genes %in% kdm2b_genes), #n23
                             sum(bcor_genes %in% kdm2b_genes), #n13
                             sum(bcor_genes[bcor_genes %in% rab_genes] %in% kdm2b_genes), #n123
                             category=c("BCOR-R","RNF2-R","KDM2B KO"),
                             cat.col=c("#A11DAB",cbPalette[1],cbPalette[5]),
                             fill=c("#A11DAB",cbPalette[1],cbPalette[5]),
                             scaled=T,euler.d=T)


bcor_genes<-as.data.frame(bcor_genes)
bcor_genes$hgnc <- hgnc[match(bcor_genes$bcor_genes,hgnc$ensembl_gene_id),"hgnc_symbol"]
write.csv(bcor_genes,file="bcor_genes.csv",quote=F, col.names = F, row.names = F)

rab_genes<-as.data.frame(rab_genes)
rab_genes$hgnc <- hgnc[match(rab_genes$rab_genes,hgnc$ensembl_gene_id),"hgnc_symbol"]
write.csv(rab_genes,file="rab_genes.csv",quote=F, col.names = F, row.names = F)

kdm2b_genes<-as.data.frame(kdm2b_genes)
kdm2b_genes$hgnc <- hgnc[match(kdm2b_genes$kdm2b_genes,hgnc$ensembl_gene_id),"hgnc_symbol"]
write.csv(kdm2b_genes,file="kdm2b_genes.csv",quote=F, col.names = F, row.names = F)

"ENSG00000163508" %in% bcor_genes$bcor_genes

x<-bcor_genes[bcor_genes$bcor_genes %in% rab_genes$rab_genes,"bcor_genes"]
View(bcor_genes[!x %in% kdm2b_genes$kdm2b_genes,])
```

#heatmaps
```{r}
dim(f_diff<-f[unique(c(bcor_genes,rab_genes,kdm2b_genes)),c(1:7,16:22,25,14,15)])
table(apply(f_diff,1,function(x) sum(x > 2.5)))
dim(f_diff<-f_diff[apply(f_diff,1,function(x) sum(x > 2.5)) > 0,])
f_diff<-log2(f_diff+0.1)
#get fold change over t=0
head(f_diff)
temp<-cbind(f_diff[,2:7]-f_diff[,1],f_diff[,9:14]-f_diff[,8],f_diff[,16:17]-f_diff[,15])

#subset to genes that go up in either first 3 days of timecourse or in KDM2B KO
temp<-temp[apply(temp,1,function(x) (x[1]> 1 | x[2] > 1 | x[3] > 1 | x[7] > 1 | x[8] > 1 | x[9] > 1 | x[13] > 1 | x[14] > 1)),]
head(temp)
idx<-rownames(temp) %in% tss_classified[tss_classified$class=="strong"]$ensembl
pheatmap(temp[idx,c(1:2,7:8,13:14)],kmeans_k=15,cluster_cols=F,show_rownames = F,scale="row",col=colors)

#f_diff<-f[rownames(res_tc_dif),c(1:7,16:22)]
#res_tc_dif$num2.5 <- apply(f_diff,1,function(x) sum(x > 2.5))
table(res_tc_dif$num2.5)
dim(res_tc_dif<-res_tc_dif[res_tc_dif$num2.5 > 0,])

colors <- colorRampPalette(c("#A2AED9","snow","#B51F2E"))(200)
#fpkm_df<- log2(f[rownames(res_tc_dif),c(1:7,16:22,25,14,15)]+0.1)
myBreaksList3 <- c(seq(min(fpkm_df,na.rm=T), 0, length.out=ceiling(length(colors)/2) + 1), 
              seq(max(fpkm_df,na.rm=T)/length(colors), max(fpkm_df,na.rm=T), length.out=floor(length(colors)/2)))

pheatmap(fpkm_df,cluster_cols=F,show_rownames = F,scale="row",col=colors)

idx<-rownames(fpkm_df) %in% tss_classified[tss_classified$class=="strong"]$ensembl
pheatmap(fpkm_df[idx,],cluster_cols=F,show_rownames = F,scale="row",col=colors)

idx2<-rownames(fpkm_df) %in% tss_classified[tss_classified$class=="weak"]$ensembl
pheatmap(fpkm_df[idx2,],cluster_cols=F,show_rownames = F,scale="row",col=colors)


```

#define new up ranges for bcor_only, kdm2b only, and both
```{r}
length(bs<-tss_classified[(tss_classified$ensembl %in% bcor_genes ) & 
                   !(tss_classified$ensembl %in% kdm2b_genes) &
                   tss_classified$class=="strong"])
length(ks<-tss_classified[!(tss_classified$ensembl %in% bcor_genes ) & 
                   (tss_classified$ensembl %in% kdm2b_genes) &
                   tss_classified$class=="strong"])
length(bks<-tss_classified[(tss_classified$ensembl %in% bcor_genes ) & 
                   (tss_classified$ensembl %in% kdm2b_genes) &
                   tss_classified$class=="strong"])

length(bw<-tss_classified[(tss_classified$ensembl %in% bcor_genes ) & 
                   !(tss_classified$ensembl %in% kdm2b_genes) &
                   tss_classified$class=="weak"])
length(kw<-tss_classified[!(tss_classified$ensembl %in% bcor_genes ) & 
                   (tss_classified$ensembl %in% kdm2b_genes) &
                   tss_classified$class=="weak"])
length(bkw<-tss_classified[(tss_classified$ensembl %in% bcor_genes ) & 
                   (tss_classified$ensembl %in% kdm2b_genes) &
                   tss_classified$class=="weak"])

bcor_v_kdm2b_granges<-GRangesList(list(bs,bks,ks,bw,bkw,kw))
names(bcor_v_kdm2b_granges)<-c("bs","bks","ks","bw","bkw","kw")
lapply(bcor_v_kdm2b_granges,length)

save("bcor_v_kdm2b_granges",file="bcor_v_kdm2b_granges.rdata")
```

